// @duetds/components/lib/cjs/duet-breadcrumb.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-00f527af.js');
const themeableComponent = require('./themeable-component-83a2a0d4.js');
const keyboardUtils = require('./keyboard-utils-7248dd0a.js');
const languageUtils = require('./language-utils-e8a527e5.js');
require('./string-utils-053749ff.js');

var actionArrowRightSmall={"title":"action-arrow-right-small","tags":"action arrow right small caret","svg":"<svg fill=\"currentColor\" viewBox=\"0 0 24 24\" xmlns=\"http://www.w3.org/2000/svg\" role=\"img\"><path d=\"M19.5 12a2.3 2.3 0 0 1-.78 1.729L7.568 23.54a1.847 1.847 0 0 1-2.439-2.773l9.752-8.579a.25.25 0 0 0 0-.376L5.129 3.233A1.847 1.847 0 0 1 7.568.46l11.148 9.808A2.31 2.31 0 0 1 19.5 12z\"/></svg>"};

const duetBreadcrumbCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:list-item;width:auto}.container{display:flex;gap:8px;align-items:center}.hide{display:none}.duet-breadrumbs-arrow{display:block;align-items:center;justify-content:center;width:0.375rem;min-width:0.375rem;height:0.375rem;min-height:0.375rem}.duet-breadcrumb{display:flex;flex-direction:row;align-items:center;justify-content:flex-start;height:100%;font-size:0.875rem;font-weight:inherit;line-height:1.25;color:inherit;text-decoration:none}.duet-breadcrumb:focus{outline:0}:host(.user-is-tabbing) .duet-breadcrumb:focus{box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px #0077b3}:host(.user-is-tabbing) .duet-breadcrumb.duet-theme-turva:focus{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px #171c3a}";

const DuetBreadcrumb = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.currentCrumb = {
            fi: "disabled. Nykyinen sivu",
            sv: "disabled. Visa mer",
            en: "disabled. Nuvarande sida",
        };
        this.hasSlottedContent = false;
        this.theme = "";
        this.language = languageUtils.getLanguage();
        this.href = "";
        this.icon = "";
        this.iconSize = "x-small";
        this.accessibleLabel = undefined;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        themeableComponent.inheritGlobalTheme(this);
    }
    connectedCallback() {
        languageUtils.connectLanguageChangeObserver(this);
        keyboardUtils.subscribeTabbingChange(this);
        this.hasSlottedContent = Array.from(this.element.childNodes).length > 0;
    }
    disconnectedCallback() {
        languageUtils.disconnectLanguageChangeObserver(this);
        keyboardUtils.unsubscribeTabbingChange(this);
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        var _a;
        return (index.h(index.Host, { role: "listitem" }, index.h("div", { class: "container" }, !this.icon && (index.h("duet-icon", { class: "duet-breadrumbs-arrow", size: "auto", icon: actionArrowRightSmall.svg, margin: "none", color: "currentColor" })), index.h("a", { class: {
                "duet-breadcrumb": true,
                "duet-theme-turva": this.theme === "turva",
            }, href: this.href ? this.href : undefined, "aria-label": `${!this.href ? `${languageUtils.getLocaleString(this.currentCrumb)} ` : ""}${(_a = this.accessibleLabel) !== null && _a !== void 0 ? _a : this.element.innerHTML}`, role: "link" }, this.icon && [
            index.h("duet-icon", { size: this.iconSize, name: this.icon, margin: "none", color: "currentColor" }),
            this.hasSlottedContent && index.h("duet-spacer", { size: "x-small", direction: "horizontal" }),
        ], index.h("slot", null)))));
    }
    get element() { return index.getElement(this); }
};
DuetBreadcrumb.style = duetBreadcrumbCss;

exports.duet_breadcrumb = DuetBreadcrumb;
