// @duetds/components/lib/cjs/duet-show-more.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-00f527af.js');
const themeableComponent = require('./themeable-component-83a2a0d4.js');
const createId = require('./create-id-1f8c438f.js');
const keyboardUtils = require('./keyboard-utils-7248dd0a.js');
const languageUtils = require('./language-utils-e8a527e5.js');
require('./string-utils-053749ff.js');

const duetShowMoreCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block}.duet-show-more{margin-bottom:16px !important;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:1rem;font-weight:600;line-height:1.5;color:rgb(0, 41, 77)}.duet-show-more.duet-m-0{margin:0 !important}.duet-show-more .slotted-content{display:none}.duet-show-more.open .slotted-content{display:block}.duet-show-more.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}";

const DuetShowMore = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.duetToggle = index.createEvent(this, "duetToggle", 7);
        this.contentId = createId.createID("DuetShowMore");
        this.buttonId = createId.createID("DuetShowMoreButton");
        this.handleKeyDown = (e) => {
            if (keyboardUtils.isKeyboardClick(e)) {
                this.handleClick(e);
            }
        };
        this.handleClick = (e) => {
            e.preventDefault();
            const open = !this.open;
            this.open = open;
            if (open) {
                setTimeout(() => {
                    this.contentElement.focus();
                }, 100);
                setTimeout(() => {
                    this.buttonAriaOpen = open;
                }, 500);
            }
            else {
                this.buttonAriaOpen = open;
            }
            this.duetToggle.emit({
                component: "duet-show-more",
                originalEvent: e,
            });
        };
        this.buttonAriaOpen = false;
        this.theme = "";
        this.language = languageUtils.getLanguage();
        this.showMoreLabelDefaults = {
            fi: "Näytä lisää",
            sv: "Visa mer",
            en: "Show more",
        };
        this.showMoreLabel = languageUtils.getLocaleString(this.showMoreLabelDefaults);
        this.showLessLabelDefaults = {
            fi: "Näytä vähemmän",
            sv: "Visa mindre",
            en: "Show less",
        };
        this.showLessLabel = languageUtils.getLocaleString(this.showLessLabelDefaults);
        this.showMoreAccessibleLabelDefaults = {
            fi: "Näytä lisää tästä aiheesta",
            sv: "Visa mer om detta ämne",
            en: "Show more about this subject",
        };
        this.showMoreAccessibleLabel = languageUtils.getLocaleString(this.showMoreAccessibleLabelDefaults);
        this.showLessAccessibleLabelDefaults = {
            fi: "Näytä vähemmän tästä aiheesta",
            sv: "Visa mindre om detta ämne",
            en: "Show less about this subject",
        };
        this.showLessAccessibleLabel = languageUtils.getLocaleString(this.showLessAccessibleLabelDefaults);
        this.margin = "auto";
        this.buttonMargin = "auto";
        this.buttonPadding = "auto";
        this.open = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        themeableComponent.inheritGlobalTheme(this);
        this.buttonAriaOpen = this.open;
    }
    connectedCallback() {
        languageUtils.connectLanguageChangeObserver(this, [
            { prop: "showMoreLabel", defaults: "showMoreLabelDefaults" },
            { prop: "showLessLabel", defaults: "showLessLabelDefaults" },
            { prop: "showMoreAccessibleLabel", defaults: "showMoreAccessibleLabelDefaults" },
            { prop: "showLessAccessibleLabel", defaults: "showLessAccessibleLabelDefaults" },
        ]);
    }
    disconnectedCallback() {
        languageUtils.disconnectLanguageChangeObserver(this);
    }
    /**
     * render() function
     * Always the last one of the class.
     */
    render() {
        return (index.h(index.Host, null, index.h("div", { class: {
                "duet-show-more": true,
                "duet-theme-turva": this.theme === "turva",
                "duet-m-0": this.margin === "none",
                open: this.open,
            } }, index.h("div", { ref: ref => (this.contentElement = ref), class: { content: true }, id: this.contentId, tabindex: this.open ? "0" : "" }, index.h("div", { class: { "slotted-content": true } }, index.h("slot", null))), index.h("duet-button", { id: this.buttonId, variation: "plain", "accessible-label": this.buttonAriaOpen ? this.showLessAccessibleLabel : this.showMoreAccessibleLabel, "accessible-expanded": this.buttonAriaOpen ? "true" : "false", "accessible-controls": this.contentId, onClick: this.handleClick, onKeyDown: this.handleKeyDown, "icon-right": true, icon: this.open ? "action-arrow-up-small" : "action-arrow-down-small", "icon-size": "medium-small", size: "small", margin: this.buttonMargin, padding: this.buttonPadding, theme: this.theme === "turva" ? "turva" : "default" }, this.open ? this.showLessLabel : this.showMoreLabel))));
    }
    get element() { return index.getElement(this); }
};
DuetShowMore.style = duetShowMoreCss;

exports.duet_show_more = DuetShowMore;
