// @duetds/components/lib/cjs/duet-file-chooser.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-00f527af.js');
const themeableComponent = require('./themeable-component-83a2a0d4.js');

const duetFileChooserCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.container{display:flex;flex-direction:row;gap:8px;align-items:center}.container .label{margin-bottom:8px}";

const DuetUploadItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.duetChange = index.createEvent(this, "duetChange", 3);
        this.onChange = (ev) => {
            ev.preventDefault();
            ev.stopPropagation();
            this.duetChange.emit({
                originalEvent: ev,
                data: { files: this.nativeInput.files },
                component: "duet-file-chooser",
            });
        };
        this.theme = "";
        this.multiple = false;
        this.accept = undefined;
    }
    /**
     * Opens file choose dialog.
     */
    async open() {
        var _a;
        (_a = this.nativeInput) === null || _a === void 0 ? void 0 : _a.click();
    }
    /**
     * Returns actually selected files in file chooser.
     */
    async getFiles() {
        return this.nativeInput.files;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        themeableComponent.inheritGlobalTheme(this);
    }
    render() {
        return (index.h(index.Host, null, index.h("duet-visually-hidden", { "aria-hidden": "true" }, index.h("input", { ref: input => {
                this.nativeInput = input;
            }, type: "file", class: {
                "duet-upload": true,
            }, "aria-hidden": "true", tabIndex: -1, onChange: this.onChange, multiple: this.multiple, accept: this.accept }))));
    }
};
DuetUploadItem.style = duetFileChooserCss;

exports.duet_file_chooser = DuetUploadItem;
