// @duetds/components/lib/cjs/duet-action-button.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-00f527af.js');
const themeableComponent = require('./themeable-component-83a2a0d4.js');
const createId = require('./create-id-1f8c438f.js');
const jsUtils = require('./js-utils-92e24ba1.js');
const keyboardUtils = require('./keyboard-utils-7248dd0a.js');
const languageUtils = require('./language-utils-e8a527e5.js');
const tokenUtils = require('./token-utils-a5356b00.js');
require('./string-utils-053749ff.js');
require('./tokens.module-ce5f9579.js');

const duetActionButtonCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:inline-block;margin:0 auto}.duet-action-button{display:inline-flex;align-items:center;justify-content:center;width:32px;height:32px;min-height:auto !important;padding:0 !important;margin:0 0 0 8px;color:rgb(0, 41, 77);white-space:nowrap;cursor:pointer;background:rgb(245, 248, 250);border-radius:50%;transition:background-color 300ms ease !important}.duet-action-button.duet-theme-turva{color:rgb(23, 28, 58);background:rgb(245, 245, 247)}.duet-action-button duet-icon{display:block}@media (max-width: 35.9375em){.duet-action-button{width:32px;height:32px}}.duet-action-button:not(:disabled):hover{background:#e9ecee}.duet-action-button:not(:disabled):hover.duet-theme-turva{background:#e9e9eb}.duet-action-button:focus{outline:0;box-shadow:0 0 0 2px #0077b3 !important}.duet-action-button:focus.duet-theme-turva{box-shadow:0 0 0 2px #171c3a !important}.duet-action-button:active:focus{background:#dddfe1 !important;box-shadow:none}.duet-action-button:active:focus.duet-theme-turva{background:#ddddde !important}.duet-action-button:disabled{color:rgb(144, 149, 153);cursor:default;background:rgb(245, 248, 250);opacity:0.75}.duet-action-button.duet-theme-turva:disabled{color:rgb(116, 116, 117);background:rgb(245, 245, 247)}";

const DuetActionButton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.duetActionEvent = index.createEvent(this, "duetActionEvent", 6);
        this.listId = createId.createID("duet-action-button");
        /**
         * Private methods.
         */
        //helper to emit relevant page events
        this.emitPageEvent = jsUtils.debounce(e => {
            this.duetActionEvent.emit({
                component: "duet-action-button",
                originalEvent: e,
                action: this.actionId,
                name: this.actionName,
                meta: this.actionMeta,
                id: this.listId,
            });
        });
        this.numbersStore = "";
        this.theme = "";
        this.iconName = "action-arrow-left-small";
        this.iconColor = "currentColor";
        this.iconSize = "xx-small";
        this.actionName = "default-action";
        this.actionId = undefined;
        this.actionMeta = undefined;
        this.disabled = false;
        this.color = "";
        this.background = "";
        this.accessibleTitle = true;
        this.accessibleDescribedBy = undefined;
        this.accessibleDetails = undefined;
        this.accessibleLabelledBy = undefined;
        this.accessibleDescription = undefined;
        this.accessibleLabelsDefaults = {
            en: "Click to activate action",
            fi: "Klikkaa aktivoidaksesi toiminto",
            sv: "Klicka för att aktivera åtgärd",
        };
        this.accessibleLabel = languageUtils.getLocaleString(this.accessibleLabelsDefaults);
        this.accessibleControls = "";
        this.url = undefined;
    }
    //handle a11y keyboard navigation events
    async onKeyboardDown(e) {
        if (keyboardUtils.isKeyboardClick(e)) {
            this.emitPageEvent(e);
        }
    }
    //return readable title for the action-button
    getTitle() {
        let titleString = "";
        if (typeof this.accessibleTitle === "string") {
            titleString = this.accessibleTitle;
        }
        if (this.accessibleTitle === true) {
            titleString = tokenUtils.capitalizeFirstLetter(`${this.actionName}`);
        }
        return titleString;
    }
    async onMouseHandler(e) {
        e.preventDefault();
        this.duetActionEvent.emit({
            component: "duet-action-button",
            originalEvent: e,
            action: this.actionName,
            id: this.actionId,
            meta: this.actionMeta,
        });
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        themeableComponent.inheritGlobalTheme(this);
    }
    connectedCallback() {
        languageUtils.connectLanguageChangeObserver(this, { prop: "accessibleLabel", defaults: "accessibleLabelsDefaults" });
    }
    disconnectedCallback() {
        languageUtils.disconnectLanguageChangeObserver(this);
    }
    /**
     * Sets focus on the specified `duet-action-button`. Use this method instead of the global
     * `button.focus()`.
     */
    async setFocus(options) {
        if (this.nativeButton) {
            this.nativeButton.focus(options);
        }
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        const color = this.color === "currentColor" ? "currentColor" : tokenUtils.getColorByName(this.color, this.theme);
        const backgroundColor = tokenUtils.getColorByName(this.background, this.theme);
        const backgroundHoverColor = backgroundColor ? tokenUtils.shadeRGBColor(backgroundColor, -0.1) : backgroundColor;
        const backgroundStyles = backgroundColor
            ? `
          button.duet-action-button {
            background: ${backgroundColor} !important;
          }
          button.duet-action-button:not(:disabled):hover {
            background: ${backgroundHoverColor} !important;
          }
        `
            : "";
        const styles = {
            color,
            "border-color": tokenUtils.getColorByName(this.iconColor, this.theme),
            "border-width": "1px",
            "border-style": "solid",
        };
        const commonProps = {
            ref: button => (this.nativeButton = button),
            part: "button",
            style: styles,
            disabled: this.disabled,
            id: this.listId,
            type: "button",
            "aria-label": this.accessibleLabel,
            "aria-labelledby": this.accessibleLabelledBy,
            "aria-description": this.accessibleDescription,
            "aria-describedby": this.accessibleDescribedBy,
            "aria-details": this.accessibleDetails,
            "aria-controls": this.accessibleControls,
            title: this.getTitle(),
            class: {
                "duet-action-button": true,
                "duet-action-button--disabled": this.disabled,
                [`duet-action-button--${this.theme}`]: this.theme !== "",
                "duet-theme-turva": this.theme === "turva",
            },
        };
        return (index.h(index.Host, { onKeyDown: e => this.onKeyboardDown(e) }, index.h("style", { innerHTML: backgroundStyles }), this.url && !this.disabled ? (index.h("a", Object.assign({}, commonProps, { href: this.url }), index.h("duet-icon", { part: "icon", name: this.iconName, color: this.iconColor, margin: "none", size: this.iconSize }))) : (index.h("button", Object.assign({}, commonProps, { onClick: e => this.onMouseHandler(e) }), index.h("duet-icon", { part: "icon", name: this.iconName, color: this.iconColor, margin: "none", size: this.iconSize }), index.h("duet-visually-hidden", null, index.h("slot", null))))));
    }
    get element() { return index.getElement(this); }
};
DuetActionButton.style = duetActionButtonCss;

exports.duet_action_button = DuetActionButton;
