// @duetds/components/lib/dist-custom-elements/duet-toggle.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h, d as Host } from './p-471a62cc.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { c as createID } from './p-5e59e970.js';
import { s as subscribeTabbingChange, u as unsubscribeTabbingChange } from './p-8f253845.js';
import { d as defineCustomElement$2 } from './p-3d6047c0.js';

const duetToggleCss = "*.sc-duet-toggle,*.sc-duet-toggle::after,*.sc-duet-toggle::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.sc-duet-toggle-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;margin-right:8px !important;margin-bottom:12px !important;-webkit-user-select:none;user-select:none;display:inline-flex;width:100%;vertical-align:bottom;-webkit-tap-highlight-color:transparent}.sc-duet-toggle-h:last-child,.sc-duet-toggle-h:last-of-type{margin-right:0 !important}.duet-m-0.sc-duet-toggle-h{margin:0 !important}duet-label.sc-duet-toggle{display:block !important;width:100% !important}.duet-side-label.sc-duet-toggle{display:flex;gap:12px;align-items:center}.duet-side-label.sc-duet-toggle duet-label.sc-duet-toggle{width:auto !important}.duet-reverse-label.sc-duet-toggle{flex-direction:row-reverse}.duet-switch.sc-duet-toggle{position:relative;display:block;width:48px;height:28px}.duet-switch.sc-duet-toggle input.sc-duet-toggle{position:absolute !important;top:0;width:1px !important;height:1px !important;padding:0 !important;overflow:hidden !important;clip:rect(1px, 1px, 1px, 1px) !important;border:0 !important}.duet-switch.sc-duet-toggle input.sc-duet-toggle:checked+.duet-slider.sc-duet-toggle{background-color:rgb(0, 119, 179);box-shadow:none}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:checked+.duet-slider.sc-duet-toggle{background-color:rgb(198, 12, 48)}.duet-switch.sc-duet-toggle input.sc-duet-toggle:checked+.duet-slider.sc-duet-toggle::before{background-color:rgb(255, 255, 255);box-shadow:-1px 0 2px #004467;transform:translateX(20px)}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:checked+.duet-slider.sc-duet-toggle::before{background-color:rgb(255, 255, 255);box-shadow:-1px 0 2px #660619}.duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled+.duet-slider.sc-duet-toggle{cursor:default}.duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled+.duet-slider.sc-duet-toggle::before{box-shadow:none}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled+.duet-slider.sc-duet-toggle::before{box-shadow:none}.duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:not(:checked)+.duet-slider.sc-duet-toggle{background:#c8cacc}.duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:not(:checked)+.duet-slider.sc-duet-toggle::before{background:rgb(255, 255, 255);box-shadow:none}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:not(:checked)+.duet-slider.sc-duet-toggle{background:#bababa}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:not(:checked)+.duet-slider.sc-duet-toggle::before{background:rgb(255, 255, 255)}.duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:checked+.duet-slider.sc-duet-toggle{background:#80bbd9}.duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:disabled:checked+.duet-slider.sc-duet-toggle{background:#e38698}.user-is-tabbing.sc-duet-toggle-h .duet-switch.sc-duet-toggle input.sc-duet-toggle:focus+.duet-slider.sc-duet-toggle{box-shadow:0 0 0 1px #fff, 0 0 0 3px rgba(0, 119, 179, 0.75)}.user-is-tabbing.sc-duet-toggle-h .duet-theme-turva.sc-duet-toggle .duet-switch.sc-duet-toggle input.sc-duet-toggle:focus+.duet-slider.sc-duet-toggle{box-shadow:0 0 0 1px #fff, 0 0 0 3px rgba(23, 28, 58, 0.75)}.duet-slider.sc-duet-toggle{position:absolute;top:0;right:0;bottom:0;left:0;cursor:pointer;background-color:rgb(144, 149, 153);border-radius:24px;transition:background-color 300ms ease}.duet-slider.sc-duet-toggle::before{position:absolute;bottom:2px;left:2px;width:24px;height:24px;content:\"\";background-color:rgb(255, 255, 255);border-radius:50%;box-shadow:1px 0 2px 0 #5d6266;transition:300ms ease}.duet-theme-turva.sc-duet-toggle .duet-slider.sc-duet-toggle{background-color:rgb(116, 116, 117)}.duet-theme-turva.sc-duet-toggle .duet-slider.sc-duet-toggle::before{background-color:rgb(255, 255, 255);box-shadow:1px 0 2px 0 #00334d}";

const DuetToggle$1 = /*@__PURE__*/ proxyCustomElement(class DuetToggle extends H {
    constructor() {
        super();
        this.__registerHost();
        this.duetChange = createEvent(this, "duetChange", 3);
        /**
         * Own Properties
         */
        this.toggleId = createID("DuetToggle");
        /**
         * Component event handling.
         */
        this.onClick = (ev) => {
            ev.preventDefault();
            this.setFocus();
            this.checked = !this.checked;
        };
        this.theme = "";
        this.margin = "auto";
        this.accessibleActiveDescendant = undefined;
        this.accessibleControls = undefined;
        this.accessibleOwns = undefined;
        this.accessibleDescribedBy = undefined;
        this.accessibleDetails = undefined;
        this.accessibleLabelledBy = undefined;
        this.accessibleDescription = undefined;
        this.label = "label";
        this.labelColor = undefined;
        this.labelWeight = undefined;
        this.labelPosition = "top";
        this.required = false;
        this.disabled = false;
        this.checked = false;
        this.value = undefined;
        this.identifier = undefined;
        this.name = undefined;
        this.role = undefined;
    }
    checkedChanged(isChecked) {
        this.duetChange.emit({
            checked: isChecked,
            value: this.value,
            component: "duet-toggle",
        });
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
        if (this.value === undefined) {
            this.value = this.identifier;
        }
    }
    connectedCallback() {
        subscribeTabbingChange(this);
    }
    disconnectedCallback() {
        unsubscribeTabbingChange(this);
    }
    /**
     * Sets focus on the specified `duet-toggle`. Use this method instead of the global
     * `input.focus()`.
     */
    async setFocus(options) {
        this.nativeInput.focus(options);
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        const identifier = this.identifier || this.toggleId;
        return (h(Host, { onClick: this.disabled ? undefined : this.onClick, class: { "duet-m-0": this.margin === "none" } }, h("div", { class: {
                "duet-toggle": true,
                "duet-theme-turva": this.theme === "turva",
                "duet-side-label": this.labelPosition !== "top",
                "duet-reverse-label": this.labelPosition === "right",
            } }, h("duet-label", { theme: this.theme, color: this.labelColor, weight: this.labelWeight, for: identifier, margin: this.labelPosition === "top" ? "auto" : "none" }, this.label), h("label", { class: "duet-switch", htmlFor: identifier }, h("input", { type: "checkbox", value: this.value, "aria-controls": this.accessibleControls, "aria-activedescendant": this.accessibleActiveDescendant, "aria-owns": this.accessibleOwns, "aria-describedby": this.accessibleDescribedBy, "aria-labelledby": this.accessibleLabelledBy, "aria-details": this.accessibleDetails, "aria-description": this.accessibleDescription, id: identifier, role: this.role, name: this.name, checked: this.checked, required: this.required, disabled: this.disabled, ref: input => (this.nativeInput = input) }), h("div", { class: "duet-slider" })))));
    }
    get element() { return this; }
    static get watchers() { return {
        "checked": ["checkedChanged"]
    }; }
    static get style() { return duetToggleCss; }
}, [2, "duet-toggle", {
        "theme": [1025],
        "margin": [1],
        "accessibleActiveDescendant": [1, "accessible-active-descendant"],
        "accessibleControls": [1, "accessible-controls"],
        "accessibleOwns": [1, "accessible-owns"],
        "accessibleDescribedBy": [1, "accessible-described-by"],
        "accessibleDetails": [1, "accessible-details"],
        "accessibleLabelledBy": [1, "accessible-labelled-by"],
        "accessibleDescription": [1, "accessible-description"],
        "label": [1],
        "labelColor": [1025, "label-color"],
        "labelWeight": [1025, "label-weight"],
        "labelPosition": [1, "label-position"],
        "required": [4],
        "disabled": [516],
        "checked": [1540],
        "value": [1537],
        "identifier": [1],
        "name": [1],
        "role": [1],
        "setFocus": [64]
    }, undefined, {
        "checked": ["checkedChanged"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-toggle", "duet-label"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-toggle":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetToggle$1);
            }
            break;
        case "duet-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const DuetToggle = DuetToggle$1;
const defineCustomElement = defineCustomElement$1;

export { DuetToggle, defineCustomElement };
