// @duetds/components/lib/dist-custom-elements/duet-pagination.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h, F as Fragment, d as Host } from './p-471a62cc.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { c as createID } from './p-5e59e970.js';
import { d as debounce } from './p-9e371678.js';
import { l as isArrowKey, m as isNumber, h as isArrowLeftKey, e as isArrowUpKey, j as isArrowRightKey, f as isArrowDownKey } from './p-8f253845.js';
import { g as getLocaleString, c as connectLanguageChangeObserver, d as disconnectLanguageChangeObserver } from './p-2e36d5bf.js';
import { p as parsePossibleJSON } from './p-0e052642.js';
import { a as breakpointTokensUpTo, b as breakpointToToken } from './p-e078836c.js';
import { d as defineCustomElement$d } from './p-4d3e1b52.js';
import { d as defineCustomElement$c } from './p-1c7bd169.js';
import { d as defineCustomElement$b } from './p-6b977095.js';
import { d as defineCustomElement$a } from './p-4c5751cd.js';
import { d as defineCustomElement$9 } from './p-c493eecb.js';
import { d as defineCustomElement$8 } from './p-3d6047c0.js';
import { d as defineCustomElement$7 } from './p-7d7291bf.js';
import { d as defineCustomElement$6 } from './p-9b174e03.js';
import { d as defineCustomElement$5 } from './p-a4605d27.js';
import { d as defineCustomElement$4 } from './p-da8acaf1.js';
import { d as defineCustomElement$3 } from './p-7bcdcf71.js';
import { d as defineCustomElement$2 } from './p-475baac7.js';

const shouldDisplayNavigation = (visibleItems, total, take) => {
    if (visibleItems * 2 > total / take) {
        return false;
    }
    return true;
};

const duetPaginationCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:inline-block;margin:0 auto}.duet-pagination-nav{position:relative}.duet-pagination ::part(button){border:0}.duet-pagination,.duet-pagination li{display:flex;flex-wrap:wrap;gap:0.1rem;align-items:center}.duet-pagination-item-is-active .duet-pagination-page-number{border-bottom-color:inherit;border-bottom-style:solid;border-bottom-width:2px}.duet-pagination duet-button{width:30px;min-width:36px;font-weight:600;text-decoration:none;border-left-width:0}.duet-pagination-overlay{position:absolute;top:1rem;z-index:300;width:100%;text-align:center}.duet-pagination.duet-pagination-is-dimmed{opacity:0.5}.duet-pagination .duet-pagination-select{margin:0 4px 0 12px !important}.duet-pagination-total-pages{margin:0;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:1rem;font-weight:600 !important;line-height:1.25;color:rgb(0, 41, 77);text-align:left}:host(.duet-theme-turva) .duet-pagination-total-pages{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}";

const DuetPagination$1 = /*@__PURE__*/ proxyCustomElement(class DuetPagination extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.duetPageChange = createEvent(this, "duetPageChange", 3);
        this.listId = createID("duet-pagination-list");
        // not in state because we don't want to trigger a render
        this.firstRenderDone = false;
        this.handleMediaQueryChange = (mq) => {
            this.matchesCompactBreakpoint = breakpointTokensUpTo.includes(this.compactBreakpoint) ? mq.matches : !mq.matches;
        };
        // called from the watcher and used to make sure any changed to current also updates the internal section index
        this.actionChain = async (after, before) => {
            if (after === 0) {
                this.current = 1;
            }
            else if (after > this.totalPages) {
                this.current = this.totalPages;
            }
            else if (before !== after) {
                const calculatedSectionIndex = Math.ceil(after / this.visibleItems) - 1;
                if (calculatedSectionIndex !== this.internalSectionIndex) {
                    this.internalSectionIndex = calculatedSectionIndex;
                    await this.setFocus();
                }
            }
            return true;
        };
        //helper to emit relevant page events
        this.emitPageEvent = (ev, passedPage) => {
            const page = Number(passedPage || this.current);
            this.duetPageChange.emit({
                component: "duet-pagination",
                from: page * this.take,
                to: page * this.take + this.take,
                current: page,
                take: this.take,
                type: "page",
                originalEvent: ev,
            });
        };
        // this will return the entered number, or the current page if the user presses enter
        this.getEnteredNumber = debounce(ev => {
            this.emitPageEvent(ev, this.numbersStore);
            this.current = Number(this.numbersStore);
            this.numbersStore = "";
            this.duetPageChange.emit({
                component: "duet-pagination",
                from: this.current * this.take,
                to: this.current * this.take + this.take,
                current: this.current,
                take: this.take,
                type: "jump",
                originalEvent: ev,
            });
        }, 500);
        this.processedPageLinks = null;
        this.numbersStore = "";
        this.matchesCompactBreakpoint = false;
        this.totalPages = 10;
        this.internalSectionIndex = 0;
        this.theme = "";
        this.variation = "default";
        this.accessibleDescribedBy = undefined;
        this.accessibleDetails = undefined;
        this.accessibleLabelledBy = undefined;
        this.accessibleDescription = undefined;
        this.accessibleLabelsDefaults = {
            en: {
                next_page: "Next page",
                previous_page: "Previous page",
                first_page: "First page",
                last_page: "Last page",
                jump_to: "Jump to page",
                go_to: "Go to page",
                description: "You may use arrow keys to select page number",
                pagination_label: "Pagination, choose a number to jump to a page",
                select_label: "Total pages {0}, choose a number to jump to a page",
                prev_section: "Show previous {0} page numbers, screen reader user should use arrowkeys to navigate",
                next_section: "Show next {0} page numbers, screen reader user should use arrowkeys to navigate",
            },
            fi: {
                next_page: "Seuraava sivu",
                previous_page: "Edellinen sivu",
                first_page: "Ensimmäinen sivu",
                last_page: "Viimeinen sivu",
                jump_to: "Siirry sivulle",
                go_to: "Siirry sivulle",
                description: "Voit käyttää nuolinäppäimiä sivunumeron valitsemiseen",
                pagination_label: "Sivutus, valitse sivunumero siirtyäksesi sivulle",
                select_label: "Sivuja yhteensä {0}, valitse sivunumero siirtyäksesi sivulle",
                prev_section: "Näytä edelliset {0} sivunumerot, käyttäjän pitää käyttää nuolinäppäimiä navigoidaksesi",
                next_section: "Näytä seuraavat {0} sivunumerot, käyttäjän pitää käyttää nuolinäppäimiä navigoidaksesi",
            },
            sv: {
                next_page: "Nästa sida",
                previous_page: "Föregående sida",
                first_page: "Första sidan",
                last_page: "Sista sidan",
                jump_to: "Hoppa till sida",
                go_to: "Hoppa till sida",
                description: "Du kan använda piltangenterna för att välja sida",
                pagination_label: "Paginering, välj en siffra för att hoppa till en sida",
                select_label: "Totalt antal sidor {0}, välj en siffra för att hoppa till en sida",
                prev_section: "Visa föregående {0} sidnummer, skrivbordsläsare ska använda piltangenterna för att navigera",
                next_section: "Visa nästa {0} sidnummer, skrivbordsläsare ska använda piltangenterna för att navigera",
            },
        };
        this.accessibleLabels = getLocaleString(this.accessibleLabelsDefaults);
        this.take = 5;
        this.visibleItems = 5;
        this.total = 1000;
        this.compactBreakpoint = "x-small";
        this.ariaControls = "";
        this.current = 1;
        this.pageLinks = undefined;
    }
    watchPropTotalHandler( /*newValue: boolean, oldValue: boolean*/) {
        this.calculatePageTake();
        return true;
    }
    watchStateHandler(newValue, oldValue) {
        return this.actionChain(newValue, oldValue);
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        // we want to keep the internal current index in sync with the current prop, but take advantage of the multiple with 0 for the first iteration
        this.internalSectionIndex = Math.ceil(this.current / this.visibleItems) - 1;
        this.calculatePageTake();
        inheritGlobalTheme(this);
        this.refresh();
    }
    componentDidLoad() {
        this.firstRenderDone = true;
    }
    componentWillRender() {
        if (this.internalSectionIndex !== 0 && this.internalSectionIndex !== this.totalPages) {
            this.calculatePageTake();
        }
    }
    componentDidRender() {
        if (this.firstRenderDone && !this.processedPageLinks) {
            this.setFocus();
        }
    }
    connectedCallback() {
        const token = this.compactBreakpoint === "none" ? "all" : breakpointToToken[this.compactBreakpoint];
        this.mq = matchMedia(token);
        this.mq.addEventListener("change", debounce(this.handleMediaQueryChange, 50));
        this.handleMediaQueryChange(this.mq);
        connectLanguageChangeObserver(this, { prop: "accessibleLabels", defaults: "accessibleLabelsDefaults" });
    }
    disconnectedCallback() {
        this.mq.removeEventListener("change", this.handleMediaQueryChange);
        this.mq = null;
        this.matchesCompactBreakpoint = false;
        disconnectLanguageChangeObserver(this);
    }
    /**
     * Private methods.
     */
    async refresh() {
        this.processedPageLinks = parsePossibleJSON(this.pageLinks);
    }
    //get totalpages from visible and total
    calculatePageTake() {
        this.totalPages = Math.ceil(this.total / this.take);
    }
    //set update current when mouse click on page number
    async onMouseHandler(e, dir) {
        if (!this.processedPageLinks) {
            e.preventDefault();
            e.stopPropagation();
            this.current = this.current + dir;
        }
        this.emitPageEvent(e);
    }
    //handle a11y keyboard navigation events
    async onKeyboardDown(e) {
        let next = this.current;
        // handle a11y based keyboard navigation for left/right/up/down
        if (isArrowLeftKey(e) || isArrowUpKey(e)) {
            next--;
        }
        if (isArrowRightKey(e) || isArrowDownKey(e)) {
            next++;
        }
        if (isArrowKey(e)) {
            if (next <= this.totalPages) {
                this.current = next;
                if (this.processedPageLinks) {
                    this.emitPageEvent(e, next - 1);
                    window.open(this.processedPageLinks[next - 1], "_self", "");
                }
                else {
                    this.emitPageEvent(e);
                }
            }
        }
        //handle the jump feature, where numbers can be entered
        // this isn't supported in NVDA where the number buttons are used by NVDA itself
        // and because of that the feature is not announced for screenreaders
        if (isNumber(e) && this.shouldDisplayNavigation()) {
            this.numbersStore = this.numbersStore + e.key;
            if (Number(this.numbersStore) > this.totalPages) {
                this.numbersStore = this.totalPages + "";
            }
            this.getEnteredNumber(e);
        }
    }
    // utility to set focus on active button
    async setFocus() {
        const currentDomElement = this.nativeNav.querySelector(`[data-id='${this.current}']`);
        currentDomElement && (await currentDomElement.setFocus());
    }
    //handles click on page numbers
    pageClickHandler(e) {
        const clickedId = Number(e.target.getAttribute("data-id"));
        if (!this.processedPageLinks) {
            e.preventDefault();
            this.current = clickedId;
        }
        this.emitPageEvent(e, clickedId);
    }
    selectClickHandler(e) {
        e.preventDefault();
        const clickedId = Number(e.target.value);
        this.current = clickedId;
        if (this.processedPageLinks) {
            this.emitPageEvent(e, clickedId - 1);
            window.open(this.processedPageLinks[clickedId - 1], "_self", "");
        }
        else {
            this.emitPageEvent(e, clickedId);
        }
    }
    //handles click on the next section buttons
    sectionClickHandler(e, direction) {
        e.preventDefault();
        let next = this.internalSectionIndex + direction;
        const takePerVisibleItems = this.take * this.visibleItems;
        if (next * takePerVisibleItems > this.total) {
            next = this.internalSectionIndex;
        }
        const newSection = next;
        this.internalSectionIndex = newSection >= 0 ? newSection : 0;
    }
    //handles click on the jump (1)button
    jumpToStart(e) {
        if (!this.processedPageLinks) {
            e.preventDefault();
            this.internalSectionIndex = 0;
            this.current = 1;
        }
        this.emitPageEvent(e);
    }
    //handles click on the jump (total) button
    jumpToEnd(e) {
        if (!this.processedPageLinks) {
            e.preventDefault();
            this.internalSectionIndex = Math.ceil(this.totalPages / this.take - 1);
            this.current = this.totalPages;
        }
        this.emitPageEvent(e);
    }
    renderPageButton(pageNumber) {
        return (h("duet-button", { class: {
                "duet-pagination-link": true,
            }, id: `page-${pageNumber}-button`, color: pageNumber !== this.current
                ? this.theme !== "turva"
                    ? "secondary"
                    : "secondary-turva"
                : this.theme !== "turva"
                    ? "primary"
                    : "primary-turva", "data-id": pageNumber, "center-text": true, variation: "plain", margin: "none", accessibleLabel: `${this.accessibleLabels.go_to} ${pageNumber}`, onClick: e => this.pageClickHandler(e), theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[pageNumber - 1] : undefined }, h("div", { class: "duet-pagination-page-number", "aria-hidden": true }, pageNumber)));
    }
    // render the individual page numbers
    renderPageNumbers() {
        const items = [];
        let i = this.internalSectionIndex * this.visibleItems;
        do {
            i++;
            if (i > this.totalPages) {
                break;
            }
            items.push(h("li", { class: {
                    "duet-pagination-item": true,
                    "duet-pagination-item-is-active": i === this.current,
                }, role: "menuitem" }, this.renderPageButton(i)));
        } while (i < this.visibleItems + this.internalSectionIndex * this.visibleItems);
        return items;
    }
    renderSelect() {
        const items = [];
        for (let i = 1; i <= this.totalPages; i++) {
            items.push({
                value: i.toString(),
                label: i.toString(),
            });
        }
        const select = (h("duet-select", { theme: this.theme, class: "duet-pagination-select", variation: "compact", items: JSON.stringify(items), value: this.current.toString(), label: this.accessibleLabels.select_label.replace("{0}", `${this.totalPages}`), labelHidden: true, onDuetChange: e => this.selectClickHandler(e) }));
        return (h("li", null, select, h("span", { class: "duet-pagination-total-pages", "aria-hidden": "true" }, "/ ", this.totalPages)));
    }
    shouldDisplayNavigation() {
        if (this.variation !== "compact" && !this.matchesCompactBreakpoint) {
            return shouldDisplayNavigation(this.visibleItems, this.total, this.take);
        }
        return false;
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        return (h(Host, { onKeyDown: e => this.onKeyboardDown(e), class: { "duet-theme-turva": this.theme === "turva" } }, h("nav", { role: "navigation", "aria-labelledby": this.accessibleLabelledBy ? `${this.accessibleLabelledBy} ${this.listId}` : this.listId, "aria-describedby": this.accessibleDescribedBy ? `${this.accessibleDescribedBy} ${this.listId}` : this.listId, "aria-descriptions": this.accessibleDescription, "aria-details": this.accessibleDetails, ref: el => (this.nativeNav = el), "aria-controls": this.ariaControls, class: {
                "duet-pagination-nav": true,
                [`duet-pagination-variation-${this.variation}`]: true,
            } }, this.numbersStore !== "" && (h("div", { "aria-live": "assertive", "aria-relevant": "text", class: "duet-pagination-overlay", "aria-label": `${this.accessibleLabels.jump_to} ${this.numbersStore}`, part: "navigation-overlay" }, h("duet-badge", { theme: this.theme }, "jump to page : ", this.numbersStore))), h("ol", { id: this.listId, class: {
                "duet-pagination": true,
                "duet-pagination-is-dimmed": this.numbersStore !== "",
            }, role: "menubar", "aria-label": this.accessibleLabels.pagination_label, "aria-description": this.accessibleLabels.description }, (this.variation === "compact" || this.matchesCompactBreakpoint) && (h("li", { class: {
                "duet-pagination-page-button": true,
                "duet-pagination-button-first": true,
            }, role: "menuitem", part: "navigation-arrow" }, h("duet-action-button", { disabled: this.internalSectionIndex === 0 && this.current <= 1, "data-direction": "back", accessibleLabel: this.accessibleLabels.first_page, title: this.accessibleLabels.first_page, onClick: e => this.jumpToStart(e), "icon-name": "action-move-to-first", theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[0] : undefined }))), h("li", { class: {
                "duet-pagination-page-button": true,
                "duet-pagination-button-first": true,
            }, role: "menuitem", part: "navigation-arrow" }, h("duet-action-button", { disabled: !!((this.internalSectionIndex === 0 && this.current <= 1) ||
                (this.current <= 1 && this.processedPageLinks)), "data-direction": "back", onClick: e => this.onMouseHandler(e, -1), "accessible-label": this.accessibleLabels.previous_page, title: this.accessibleLabels.previous_page, "icon-name": "action-arrow-left-small", theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[this.current - 2] : undefined })), this.internalSectionIndex !== 0 && this.shouldDisplayNavigation() && (h(Fragment, null, h("li", { class: "duet-pagination-item", role: "menuitem", part: "navigation-jump" }, h("duet-button", { class: "duet-pagination-link duet-pagination-link-jump-back", "center-text": "true", variation: "plain", color: this.theme !== "turva" ? "secondary" : "secondary-turva", margin: "none", accessibleLabel: this.accessibleLabels.first_page, onClick: e => this.jumpToStart(e), theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[0] : undefined }, h("div", { class: "duet-pagination-page-number" }, "1"))), h("li", { class: "duet-pagination-item", role: "menuitem", part: "navigation-section" }, h("duet-button", { class: "duet-pagination-link duet-pagination-link-dots-back", "center-text": "true", color: this.theme !== "turva" ? "secondary" : "secondary-turva", variation: "plain", margin: "none", accessibleLabel: this.accessibleLabels.prev_section.replace("{0}", `${this.visibleItems}`), onClick: e => this.sectionClickHandler(e, -1), theme: this.theme }, h("div", { class: "duet-pagination-page-number", "aria-hidden": "true" }, "\u2026"))))), (this.shouldDisplayNavigation() && this.renderPageNumbers()) || this.renderSelect(), (this.internalSectionIndex + 1) * this.visibleItems < this.totalPages &&
            this.shouldDisplayNavigation() && (h(Fragment, null, (this.internalSectionIndex + 1) * this.visibleItems !== this.totalPages - 1 && (h("li", { class: "duet-pagination-item", role: "menuitem", part: "navigation-section" }, h("duet-button", { class: "duet-pagination-link duet-pagination-link-dots-forward", "center-text": "true", color: this.theme !== "turva" ? "secondary" : "secondary-turva", variation: "plain", margin: "none", accessibleLabel: this.accessibleLabels.next_section.replace("{0}", `${this.visibleItems}`), onClick: e => this.sectionClickHandler(e, 1), theme: this.theme }, h("div", { class: "duet-pagination-page-number", "aria-hidden": "true" }, "\u2026")))), h("li", { class: "duet-pagination-item", role: "menuitem", part: "navigation-jump" }, h("duet-button", { class: "duet-pagination-link duet-pagination-link-jump-forward", "center-text": "true", color: this.theme !== "turva" ? "secondary" : "secondary-turva", variation: "plain", margin: "none", accessibleLabel: this.accessibleLabels.last_page, onClick: e => this.jumpToEnd(e), theme: this.theme, url: this.processedPageLinks
                ? this.processedPageLinks[this.processedPageLinks.length - 1]
                : undefined }, h("div", { class: "duet-pagination-page-number" }, this.totalPages))))), h("li", { class: {
                "duet-pagination-page-button": true,
                "duet-pagination-button-last": true,
            }, role: "menuitem", part: "navigation-arrow" }, h("duet-action-button", { disabled: !!((this.current === this.totalPages &&
                this.internalSectionIndex === Math.ceil(this.totalPages / this.visibleItems) - 1) ||
                (this.current === this.totalPages && this.processedPageLinks)), class: "duet-date-dialog-next", "accessible-label": this.accessibleLabels.next_page, title: this.accessibleLabels.next_page, onClick: e => this.onMouseHandler(e, 1), "icon-name": "action-move-to-next", theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[this.current] : undefined })), (this.variation === "compact" || this.matchesCompactBreakpoint) && (h("li", { class: {
                "duet-pagination-page-button": true,
                "duet-pagination-button-last": true,
            }, role: "menuitem", part: "navigation-arrow" }, h("duet-action-button", { disabled: this.current === this.totalPages &&
                this.internalSectionIndex === Math.ceil(this.totalPages / this.visibleItems) - 1, accessibleLabel: this.accessibleLabels.last_page, title: this.accessibleLabels.last_page, onClick: e => this.jumpToEnd(e), "icon-name": "action-move-to-last", theme: this.theme, url: this.processedPageLinks ? this.processedPageLinks[this.totalPages - 1] : undefined })))))));
    }
    get element() { return this; }
    static get watchers() { return {
        "total": ["watchPropTotalHandler"],
        "current": ["watchStateHandler"],
        "pageLinks": ["refresh"]
    }; }
    static get style() { return duetPaginationCss; }
}, [1, "duet-pagination", {
        "theme": [1025],
        "variation": [1],
        "accessibleDescribedBy": [1, "accessible-described-by"],
        "accessibleDetails": [1, "accessible-details"],
        "accessibleLabelledBy": [1, "accessible-labelled-by"],
        "accessibleDescription": [1, "accessible-description"],
        "accessibleLabelsDefaults": [16],
        "accessibleLabels": [1040],
        "take": [2],
        "visibleItems": [2, "visible-items"],
        "total": [2],
        "compactBreakpoint": [1, "compact-breakpoint"],
        "ariaControls": [1, "accessible-controls"],
        "current": [1026],
        "pageLinks": [1025, "page-links"],
        "processedPageLinks": [32],
        "numbersStore": [32],
        "matchesCompactBreakpoint": [32],
        "totalPages": [32],
        "internalSectionIndex": [32]
    }, undefined, {
        "total": ["watchPropTotalHandler"],
        "current": ["watchStateHandler"],
        "pageLinks": ["refresh"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-pagination", "duet-action-button", "duet-badge", "duet-button", "duet-caption", "duet-icon", "duet-label", "duet-select", "duet-spinner", "duet-tooltip", "duet-tooltip-button", "duet-tooltip-popup", "duet-visually-hidden"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-pagination":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetPagination$1);
            }
            break;
        case "duet-action-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$d();
            }
            break;
        case "duet-badge":
            if (!customElements.get(tagName)) {
                defineCustomElement$c();
            }
            break;
        case "duet-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$b();
            }
            break;
        case "duet-caption":
            if (!customElements.get(tagName)) {
                defineCustomElement$a();
            }
            break;
        case "duet-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "duet-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "duet-select":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "duet-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "duet-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "duet-tooltip-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "duet-tooltip-popup":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "duet-visually-hidden":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const DuetPagination = DuetPagination$1;
const defineCustomElement = defineCustomElement$1;

export { DuetPagination, defineCustomElement };
