// @duetds/components/lib/dist-custom-elements/duet-show-more.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h, d as Host } from './p-471a62cc.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { c as createID } from './p-5e59e970.js';
import { i as isKeyboardClick } from './p-8f253845.js';
import { a as getLanguage, g as getLocaleString, c as connectLanguageChangeObserver, d as disconnectLanguageChangeObserver } from './p-2e36d5bf.js';
import { d as defineCustomElement$5 } from './p-6b977095.js';
import { d as defineCustomElement$4 } from './p-c493eecb.js';
import { d as defineCustomElement$3 } from './p-9b174e03.js';
import { d as defineCustomElement$2 } from './p-475baac7.js';

const duetShowMoreCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block}.duet-show-more{margin-bottom:16px !important;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:1rem;font-weight:600;line-height:1.5;color:rgb(0, 41, 77)}.duet-show-more.duet-m-0{margin:0 !important}.duet-show-more .slotted-content{display:none}.duet-show-more.open .slotted-content{display:block}.duet-show-more.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}";

const DuetShowMore$1 = /*@__PURE__*/ proxyCustomElement(class DuetShowMore extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.duetToggle = createEvent(this, "duetToggle", 7);
        this.contentId = createID("DuetShowMore");
        this.buttonId = createID("DuetShowMoreButton");
        this.handleKeyDown = (e) => {
            if (isKeyboardClick(e)) {
                this.handleClick(e);
            }
        };
        this.handleClick = (e) => {
            e.preventDefault();
            const open = !this.open;
            this.open = open;
            if (open) {
                setTimeout(() => {
                    this.contentElement.focus();
                }, 100);
                setTimeout(() => {
                    this.buttonAriaOpen = open;
                }, 500);
            }
            else {
                this.buttonAriaOpen = open;
            }
            this.duetToggle.emit({
                component: "duet-show-more",
                originalEvent: e,
            });
        };
        this.buttonAriaOpen = false;
        this.theme = "";
        this.language = getLanguage();
        this.showMoreLabelDefaults = {
            fi: "Näytä lisää",
            sv: "Visa mer",
            en: "Show more",
        };
        this.showMoreLabel = getLocaleString(this.showMoreLabelDefaults);
        this.showLessLabelDefaults = {
            fi: "Näytä vähemmän",
            sv: "Visa mindre",
            en: "Show less",
        };
        this.showLessLabel = getLocaleString(this.showLessLabelDefaults);
        this.showMoreAccessibleLabelDefaults = {
            fi: "Näytä lisää tästä aiheesta",
            sv: "Visa mer om detta ämne",
            en: "Show more about this subject",
        };
        this.showMoreAccessibleLabel = getLocaleString(this.showMoreAccessibleLabelDefaults);
        this.showLessAccessibleLabelDefaults = {
            fi: "Näytä vähemmän tästä aiheesta",
            sv: "Visa mindre om detta ämne",
            en: "Show less about this subject",
        };
        this.showLessAccessibleLabel = getLocaleString(this.showLessAccessibleLabelDefaults);
        this.margin = "auto";
        this.buttonMargin = "auto";
        this.buttonPadding = "auto";
        this.open = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
        this.buttonAriaOpen = this.open;
    }
    connectedCallback() {
        connectLanguageChangeObserver(this, [
            { prop: "showMoreLabel", defaults: "showMoreLabelDefaults" },
            { prop: "showLessLabel", defaults: "showLessLabelDefaults" },
            { prop: "showMoreAccessibleLabel", defaults: "showMoreAccessibleLabelDefaults" },
            { prop: "showLessAccessibleLabel", defaults: "showLessAccessibleLabelDefaults" },
        ]);
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
    }
    /**
     * render() function
     * Always the last one of the class.
     */
    render() {
        return (h(Host, null, h("div", { class: {
                "duet-show-more": true,
                "duet-theme-turva": this.theme === "turva",
                "duet-m-0": this.margin === "none",
                open: this.open,
            } }, h("div", { ref: ref => (this.contentElement = ref), class: { content: true }, id: this.contentId, tabindex: this.open ? "0" : "" }, h("div", { class: { "slotted-content": true } }, h("slot", null))), h("duet-button", { id: this.buttonId, variation: "plain", "accessible-label": this.buttonAriaOpen ? this.showLessAccessibleLabel : this.showMoreAccessibleLabel, "accessible-expanded": this.buttonAriaOpen ? "true" : "false", "accessible-controls": this.contentId, onClick: this.handleClick, onKeyDown: this.handleKeyDown, "icon-right": true, icon: this.open ? "action-arrow-up-small" : "action-arrow-down-small", "icon-size": "medium-small", size: "small", margin: this.buttonMargin, padding: this.buttonPadding, theme: this.theme === "turva" ? "turva" : "default" }, this.open ? this.showLessLabel : this.showMoreLabel))));
    }
    get element() { return this; }
    static get style() { return duetShowMoreCss; }
}, [1, "duet-show-more", {
        "theme": [1025],
        "language": [1537],
        "showMoreLabelDefaults": [16],
        "showMoreLabel": [1537, "show-more-label"],
        "showLessLabelDefaults": [16],
        "showLessLabel": [1537, "show-less-label"],
        "showMoreAccessibleLabelDefaults": [16],
        "showMoreAccessibleLabel": [1537, "show-more-accessible-label"],
        "showLessAccessibleLabelDefaults": [16],
        "showLessAccessibleLabel": [1537, "show-less-accessible-label"],
        "margin": [1],
        "buttonMargin": [1, "button-margin"],
        "buttonPadding": [1, "button-padding"],
        "open": [1540],
        "buttonAriaOpen": [32]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-show-more", "duet-button", "duet-icon", "duet-spinner", "duet-visually-hidden"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-show-more":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetShowMore$1);
            }
            break;
        case "duet-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "duet-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "duet-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "duet-visually-hidden":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const DuetShowMore = DuetShowMore$1;
const defineCustomElement = defineCustomElement$1;

export { DuetShowMore, defineCustomElement };
