// @duetds/components/lib/dist-custom-elements/duet-submenu-bar-dropdown-link.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-471a62cc.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { b as isEscapeKey } from './p-8f253845.js';
import { d as defineCustomElement$2 } from './p-c493eecb.js';

const duetSubmenuBarDropdownLinkCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:100%}.duet-submenu-bar-dropdown-link{display:flex;flex-direction:row;gap:12px;align-items:center;justify-content:flex-start;height:100%;padding:8px 20px 8px 60px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:1rem;font-weight:400;line-height:1.25;color:rgb(0, 41, 77);text-align:left;text-decoration:none}.duet-submenu-bar-dropdown-link:focus{outline:0;box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px #0077b3}.duet-theme-turva .duet-submenu-bar-dropdown-link:focus,.duet-submenu-bar-dropdown-link:focus.duet-theme-turva{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px #171c3a}@media (min-width: 62em){.duet-submenu-bar-dropdown-link{padding:0.5rem 8px;font-size:1rem;line-height:1.5}}.duet-submenu-bar-dropdown-link.active .label{position:relative}.duet-submenu-bar-dropdown-link.active .label::after{position:absolute;top:auto;bottom:0;left:0;display:block;width:100%;height:1px;content:\"\";background:rgb(0, 119, 179);transform:1px}.duet-submenu-bar-dropdown-link.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}.duet-submenu-bar-dropdown-link.duet-theme-turva.active .label::after{background:rgb(198, 12, 48)}@media (min-width: 62em){.duet-submenu-bar-dropdown-link.active{background:rgb(243, 249, 252)}.duet-submenu-bar-dropdown-link.active .label::after{display:none}.duet-submenu-bar-dropdown-link.active.duet-theme-turva{background:rgb(245, 245, 247)}.duet-submenu-bar-dropdown-link:hover{background:rgb(230, 242, 248)}.duet-submenu-bar-dropdown-link:hover.duet-theme-turva{background:rgb(228, 228, 230)}}.duet-submenu-bar-dropdown-link.highlight{color:rgb(0, 119, 179)}.duet-submenu-bar-dropdown-link.highlight.duet-theme-turva{color:rgb(148, 9, 37)}";

const DuetSubmenuBarDropdownLink$1 = /*@__PURE__*/ proxyCustomElement(class DuetSubmenuBarDropdownLink extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.theme = "";
        this.active = false;
        this.href = "#";
        this.icon = undefined;
        this.caret = false;
        this.highlight = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    handleEscape(e) {
        if (isEscapeKey(e)) {
            const dropdown = this.element.closest("duet-submenu-bar-dropdown");
            dropdown.open = false;
            dropdown.setFocus();
        }
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        return (h(Host, null, h("a", { class: {
                "duet-submenu-bar-dropdown-link": true,
                "duet-theme-turva": this.theme === "turva",
                highlight: this.highlight,
                active: this.active,
            }, "aria-current": this.active ? "true" : "false", href: this.href, onKeyUp: e => this.handleEscape(e) }, this.icon && h("duet-icon", { name: this.icon, size: "x-small", margin: "none", color: "currentColor" }), h("span", { class: "label" }, h("slot", null)), this.caret && h("duet-icon", { name: "action-arrow-right-small", size: "xxx-small", margin: "none" }))));
    }
    get element() { return this; }
    static get style() { return duetSubmenuBarDropdownLinkCss; }
}, [1, "duet-submenu-bar-dropdown-link", {
        "theme": [1025],
        "active": [1028],
        "href": [1025],
        "icon": [1537],
        "caret": [1028],
        "highlight": [1028]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-submenu-bar-dropdown-link", "duet-icon"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-submenu-bar-dropdown-link":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetSubmenuBarDropdownLink$1);
            }
            break;
        case "duet-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const DuetSubmenuBarDropdownLink = DuetSubmenuBarDropdownLink$1;
const defineCustomElement = defineCustomElement$1;

export { DuetSubmenuBarDropdownLink, defineCustomElement };
