// @duetds/components/lib/dist-custom-elements/duet-file-chooser.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h, d as Host } from './p-471a62cc.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { d as defineCustomElement$2 } from './p-475baac7.js';

const duetFileChooserCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.container{display:flex;flex-direction:row;gap:8px;align-items:center}.container .label{margin-bottom:8px}";

const DuetUploadItem = /*@__PURE__*/ proxyCustomElement(class DuetUploadItem extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.duetChange = createEvent(this, "duetChange", 3);
        this.onChange = (ev) => {
            ev.preventDefault();
            ev.stopPropagation();
            this.duetChange.emit({
                originalEvent: ev,
                data: { files: this.nativeInput.files },
                component: "duet-file-chooser",
            });
        };
        this.theme = "";
        this.multiple = false;
        this.accept = undefined;
    }
    /**
     * Opens file choose dialog.
     */
    async open() {
        var _a;
        (_a = this.nativeInput) === null || _a === void 0 ? void 0 : _a.click();
    }
    /**
     * Returns actually selected files in file chooser.
     */
    async getFiles() {
        return this.nativeInput.files;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    render() {
        return (h(Host, null, h("duet-visually-hidden", { "aria-hidden": "true" }, h("input", { ref: input => {
                this.nativeInput = input;
            }, type: "file", class: {
                "duet-upload": true,
            }, "aria-hidden": "true", tabIndex: -1, onChange: this.onChange, multiple: this.multiple, accept: this.accept }))));
    }
    static get style() { return duetFileChooserCss; }
}, [1, "duet-file-chooser", {
        "theme": [1025],
        "multiple": [4],
        "accept": [1],
        "open": [64],
        "getFiles": [64]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-file-chooser", "duet-visually-hidden"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-file-chooser":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetUploadItem);
            }
            break;
        case "duet-visually-hidden":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const DuetFileChooser = DuetUploadItem;
const defineCustomElement = defineCustomElement$1;

export { DuetFileChooser, defineCustomElement };
