// @duetds/components/lib/dist-custom-elements/duet-popup-menu.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, c as createEvent, h, d as Host } from './p-471a62cc.js';
import { c as computePosition, o as offset, f as flip } from './p-00c8a755.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { b as isEscapeKey, f as isArrowDownKey, e as isArrowUpKey, d as isTabKey } from './p-8f253845.js';
import { T as Teleport } from './p-ccdff614.js';

const duetPopupMenuCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.duet-popup-menu-list{position:absolute;top:0;left:0;z-index:950;width:max-content;min-width:175px;padding:8px;margin:0;color:rgb(0, 41, 77);list-style:none;visibility:hidden !important;background:rgb(255, 255, 255);border-radius:8px;box-shadow:0 0 0 1px rgba(0, 0, 0, 0.07), 0 4px 10px 0 rgba(0, 41, 77, 0.15);opacity:0;transition:transform 300ms ease, visibility 300ms ease, opacity 300ms ease;transform:scaleY(0);transform-origin:top center}.duet-popup-menu-list.open{visibility:visible !important;opacity:1;transform:scaleY(1)}.duet-popup-menu-list.fixed{position:fixed}.duet-popup-menu-list.duet-theme-turva{color:rgb(23, 28, 58) !important;background:rgb(255, 255, 255)}";

const DuetPopupMenu$1 = /*@__PURE__*/ proxyCustomElement(class DuetPopupMenu extends H {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.duetToggle = createEvent(this, "duetToggle", 7);
        this.updatePlacement = async () => {
            if (!this.controller) {
                return;
            }
            const coordinates = await computePosition(this.controller, this.element, {
                middleware: [offset({ mainAxis: this.mainAxisOffset, crossAxis: this.crossAxisOffset }), flip()],
                placement: this.placement,
                strategy: this.position,
            });
            this.menuY = coordinates.y;
            this.menuX = coordinates.x;
        };
        this.open = false;
        this.menuY = undefined;
        this.menuX = undefined;
        this.accessibleLabel = undefined;
        this.theme = "";
        this.controller = undefined;
        this.placement = "bottom-start";
        this.position = "absolute";
        this.mainAxisOffset = 0;
        this.crossAxisOffset = 0;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
        this.teleport = new Teleport(this.element);
    }
    componentDidLoad() {
        {
            this.updatePlacement();
        }
    }
    /**
     * Component event handling.
     */
    handleKeyUp(evt) {
        if (this.open) {
            // When Esc is pressed the menu should be closed
            if (isEscapeKey(evt)) {
                this.focusController();
                this.hide();
            }
        }
    }
    async handleKeyDown(evt) {
        if (this.open) {
            if (isArrowDownKey(evt)) {
                evt.preventDefault();
                const current = await this.getFocusedItemIndex();
                if (current !== -1) {
                    if (current < this.itemElements.length - 1) {
                        this.itemElements[current + 1].setFocus();
                    }
                    else {
                        this.itemElements[0].setFocus();
                    }
                }
            }
            if (isArrowUpKey(evt)) {
                evt.preventDefault();
                const current = await this.getFocusedItemIndex();
                if (current !== -1) {
                    if (current === 0) {
                        this.itemElements[this.itemElements.length - 1].setFocus();
                    }
                    else {
                        this.itemElements[current - 1].setFocus();
                    }
                }
            }
            if (isTabKey(evt)) {
                evt.preventDefault();
                this.hide();
            }
        }
    }
    handleClick(evt) {
        const isClickOutside = evt.composedPath().every(node => node !== this.element && node !== this.controller);
        if (this.open && isClickOutside) {
            this.hide();
        }
    }
    /**
     * isOpen.
     */
    async isOpen() {
        return this.open;
    }
    /**
     * Show.
     */
    async show() {
        this.open = true;
        this.teleport.go();
        this.itemElements[0].setFocus();
        this.duetToggle.emit({ open: this.open, component: "duet-popup-menu" });
        this.updatePlacement();
    }
    /**
     * Hide.
     */
    async hide() {
        this.open = false;
        this.duetToggle.emit({ open: this.open, component: "duet-popup-menu" });
        this.focusController();
        this.teleport.resume(200);
    }
    /**
     * Toggle.
     */
    async toggle() {
        if (this.open) {
            this.hide();
        }
        else {
            this.show();
        }
    }
    focusController() {
        if (this.controller) {
            if ("setFocus" in this.controller) {
                this.controller.setFocus();
            }
            else {
                this.controller.focus();
            }
        }
    }
    get itemElements() {
        return Array.from(this.element.querySelectorAll(":scope > duet-popup-menu-item"));
    }
    async getFocusedItemIndex() {
        const items = this.itemElements;
        for (let i = 0; i < items.length; i++) {
            if (await items[i].isFocused()) {
                return i;
            }
        }
        return -1;
    }
    render() {
        const positionStyles = {
            top: `${this.menuY}px`,
            left: `${this.menuX}px`,
        };
        return (h(Host, { class: this.position }, h("div", { style: positionStyles, class: { open: this.open, "duet-popup-menu-list": true, "duet-theme-turva": this.theme === "turva" }, role: "menu", "aria-label": this.accessibleLabel }, h("slot", null))));
    }
    get element() { return this; }
    static get style() { return duetPopupMenuCss; }
}, [1, "duet-popup-menu", {
        "accessibleLabel": [1, "accessible-label"],
        "theme": [1025],
        "controller": [16],
        "placement": [1],
        "position": [1],
        "mainAxisOffset": [2, "main-axis-offset"],
        "crossAxisOffset": [2, "cross-axis-offset"],
        "open": [32],
        "menuY": [32],
        "menuX": [32],
        "isOpen": [64],
        "show": [64],
        "hide": [64],
        "toggle": [64]
    }, [[8, "keyup", "handleKeyUp"], [8, "keydown", "handleKeyDown"], [8, "click", "handleClick"]]]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-popup-menu"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-popup-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetPopupMenu$1);
            }
            break;
    } });
}

const DuetPopupMenu = DuetPopupMenu$1;
const defineCustomElement = defineCustomElement$1;

export { DuetPopupMenu, defineCustomElement };
