// @duetds/components/lib/dist-custom-elements/p-a4605d27.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-471a62cc.js';
import { m as media_query_medium, a as media_query_large } from './p-5e849463.js';
import { c as computePosition, f as flip, s as shift, a as arrow } from './p-00c8a755.js';
import { i as inheritGlobalTheme } from './p-a58bd561.js';
import { c as createID } from './p-5e59e970.js';
import { b as isEscapeKey } from './p-8f253845.js';
import { a as getLanguage, g as getLocaleString, c as connectLanguageChangeObserver, d as disconnectLanguageChangeObserver } from './p-2e36d5bf.js';
import { d as defineCustomElement$4 } from './p-c493eecb.js';
import { d as defineCustomElement$3 } from './p-da8acaf1.js';
import { d as defineCustomElement$2 } from './p-7bcdcf71.js';
import { d as defineCustomElement$1 } from './p-475baac7.js';

const duetTooltipCss = "*.sc-duet-tooltip,*.sc-duet-tooltip::after,*.sc-duet-tooltip::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.sc-duet-tooltip-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;z-index:200;display:inline-flex;width:auto;margin-top:-16px !important;margin-left:-3px !important;-webkit-tap-highlight-color:transparent}duet-tooltip-button.sc-duet-tooltip{position:relative}duet-tooltip-popup.sc-duet-tooltip{position:absolute}";

const DuetTooltip = /*@__PURE__*/ proxyCustomElement(class DuetTooltip extends H {
    constructor() {
        super();
        this.__registerHost();
        /**
         * Own Properties.
         */
        this.mql = [
            window.matchMedia(media_query_medium.replace(/'/g, "")),
            window.matchMedia(media_query_large.replace(/'/g, "")),
        ];
        this.tooltipId = createID("DuetTooltip");
        this.aboutString = {
            fi: "aiheesta",
            en: "about",
            sv: "om",
        };
        /**
         * Local methods
         */
        this.openTooltip = () => {
            this.active = true;
            this.focusTarget = "tooltip";
        };
        this.closeTooltip = (moveFocusToButton = false) => {
            this.active = false;
            if (moveFocusToButton) {
                this.focusTarget = "trigger";
            }
            setTimeout(() => {
                this.updatePlacement();
            }, 200);
        };
        this.toggleOpen = () => {
            this.active ? this.closeTooltip() : this.openTooltip();
        };
        this.updatePlacement = async () => {
            var _a, _b;
            let preferRight = true;
            if (this.breakpoint === "large" && !this.mql[1].matches) {
                preferRight = false;
            }
            else if (this.breakpoint === "medium" && !this.mql[0].matches) {
                preferRight = false;
            }
            let placement = preferRight ? "top-start" : "top-end";
            if (this.direction === "left") {
                placement = "top-end";
            }
            else if (this.direction === "right") {
                placement = "top-start";
            }
            const position = await computePosition(this.buttonEl, this.popupElement, {
                placement,
                middleware: [flip(), shift({ padding: 5 }), arrow({ element: this.arrowEl })],
            });
            this.tooltipY = position.y;
            this.tooltipX = position.x;
            this.placement = position.placement;
            this.arrowY = (_a = position.middlewareData.arrow) === null || _a === void 0 ? void 0 : _a.y;
            this.arrowX = (_b = position.middlewareData.arrow) === null || _b === void 0 ? void 0 : _b.x;
        };
        this.placement = "bottom";
        this.tooltipY = undefined;
        this.tooltipX = undefined;
        this.arrowX = undefined;
        this.arrowY = undefined;
        this.theme = "";
        this.language = getLanguage();
        this.accessibleLabelDefault = {
            fi: "Näytä lisätietoja",
            en: "Show more information",
            sv: "Visa mer tilläggsuppgifter",
        };
        this.accessibleLabel = getLocaleString(this.accessibleLabelDefault, this.language);
        this.accessibleAboutLabel = undefined;
        this.accessibleInputLabel = undefined;
        this.accessibleCloseLabelDefault = {
            fi: "Sulje lisätiedot",
            en: "Close details",
            sv: "Stäng tilläggsuppgifter",
        };
        this.accessibleCloseLabel = getLocaleString(this.accessibleCloseLabelDefault, this.language);
        this.label = "";
        this.breakpoint = "medium";
        this.direction = "auto";
        this.positionVertical = "0";
        this.positionHorizontal = "0";
        this.active = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    async componentDidLoad() {
        this.arrowEl = await this.popupElement.getArrowElement();
        this.contentEl = await this.popupElement.getContentElement();
        this.updatePlacement();
        if (this.active) {
            this.openTooltip();
        }
        this.intersectionObserver = new IntersectionObserver(this.updatePlacement, { root: document.body });
        this.intersectionObserver.observe(this.element);
    }
    componentDidUpdate() {
        switch (this.focusTarget) {
            case "tooltip":
                // delay focus so button is correctly announced as "expanded"
                setTimeout(() => this.contentEl.focus({ preventScroll: true }));
                break;
            case "trigger":
                // delay focus so button is correctly announced as "collapsed"
                setTimeout(() => this.buttonEl.setFocus({ preventScroll: true }));
                break;
        }
        this.focusTarget = null;
    }
    connectedCallback() {
        connectLanguageChangeObserver(this, [
            { prop: "accessibleLabel", defaults: "accessibleLabelDefault" },
            { prop: "accessibleCloseLabel", defaults: "accessibleCloseLabelDefault" },
        ]);
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
    }
    /**
     * Component event handling.
     */
    onScroll() {
        clearTimeout(this.scrollTimer);
        this.scrollTimer = setTimeout(() => {
            if (this.active) {
                this.updatePlacement();
            }
        }, 100);
    }
    handleWindowResize() {
        clearTimeout(this.resizeTimer);
        this.resizeTimer = setTimeout(() => {
            this.updatePlacement();
        }, 100);
    }
    handleDocumentClick(e) {
        if (!this.active) {
            return;
        }
        const path = e.composedPath();
        const isClickOutside = path.every(el => el !== this.element && el !== this.contentEl);
        if (isClickOutside) {
            this.closeTooltip();
        }
    }
    handleKeyDown(e) {
        if (this.active && isEscapeKey(e)) {
            this.closeTooltip(true);
        }
    }
    handlePopupClose() {
        if (this.active) {
            this.closeTooltip(true);
        }
    }
    async handleActiveChange(newActiveValue) {
        if (newActiveValue) {
            setTimeout(() => this.updatePlacement(), 10);
        }
        else {
            setTimeout(() => this.updatePlacement(), 300);
        }
    }
    /**
     * render() function.
     * Always the last one in the class.
     */
    render() {
        const tooltipStyles = {
            transform: `translateY(${this.positionVertical}) translateX(${this.positionHorizontal})`,
            // This is needed to that the active tooltip stacks on
            // top of other tooltips and the header.
            zIndex: this.active ? "201" : "",
        };
        const popupContainerStyles = {
            position: "relative",
            overflow: this.active ? "visible" : "hidden",
        };
        const positionStyles = {
            top: `${this.tooltipY}px`,
            left: `${this.tooltipX}px`,
            display: "block",
            "pointer-events": this.active ? "auto" : "none",
        };
        const arrowStyles = {
            top: this.arrowY ? `${this.arrowY}px` : undefined,
            left: this.arrowX ? `${this.arrowX}px` : undefined,
        };
        return (h(Host, { style: tooltipStyles }, h("duet-tooltip-button", { ref: element => (this.buttonEl = element), active: this.active, tooltipId: this.tooltipId, label: this.label, accessibleLabel: (this.accessibleInputLabel && `${this.accessibleInputLabel}, ${this.accessibleLabel}`) ||
                (this.accessibleAboutLabel &&
                    `${this.accessibleLabel} ${this.aboutString[this.language]} ${this.accessibleAboutLabel}`) ||
                this.accessibleLabel, onClick: this.toggleOpen, theme: this.theme }), h("div", { style: popupContainerStyles }, h("duet-tooltip-popup", { ref: element => (this.popupElement = element), active: this.active, tooltipId: this.tooltipId, arrowStyles: arrowStyles, style: positionStyles, placement: this.placement, accessibleCloseLabel: this.accessibleCloseLabel, theme: this.theme }, h("slot", null)))));
    }
    get element() { return this; }
    static get watchers() { return {
        "active": ["handleActiveChange"]
    }; }
    static get style() { return duetTooltipCss; }
}, [6, "duet-tooltip", {
        "theme": [1025],
        "language": [1025],
        "accessibleLabelDefault": [1, "accessible-label-default"],
        "accessibleLabel": [1025, "accessible-label"],
        "accessibleAboutLabel": [1, "accessible-about-label"],
        "accessibleInputLabel": [1, "accessible-input-label"],
        "accessibleCloseLabelDefault": [1, "accessible-close-label-default"],
        "accessibleCloseLabel": [1025, "accessible-close-label"],
        "label": [1],
        "breakpoint": [1],
        "direction": [1],
        "positionVertical": [513, "position-vertical"],
        "positionHorizontal": [513, "position-horizontal"],
        "active": [1540],
        "placement": [32],
        "tooltipY": [32],
        "tooltipX": [32],
        "arrowX": [32],
        "arrowY": [32]
    }, [[9, "scroll", "onScroll"], [9, "resize", "handleWindowResize"], [4, "click", "handleDocumentClick"], [4, "keydown", "handleKeyDown"], [4, "tooltipPopupClosed", "handlePopupClose"]], {
        "active": ["handleActiveChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["duet-tooltip", "duet-icon", "duet-tooltip-button", "duet-tooltip-popup", "duet-visually-hidden"];
    components.forEach(tagName => { switch (tagName) {
        case "duet-tooltip":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, DuetTooltip);
            }
            break;
        case "duet-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "duet-tooltip-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "duet-tooltip-popup":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "duet-visually-hidden":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { DuetTooltip as D, defineCustomElement as d };
