// @duetds/components/lib/collection/components/duet-tab/duet-tab.js
/*!
 * Built with Duet Design System
 */
import { Build, h } from "@stencil/core";
import { connectLanguageChangeObserver, disconnectLanguageChangeObserver, getLocaleString, } from "../../utils/language-utils";
export class DuetTab {
    constructor() {
        this.label = undefined;
        this.caption = undefined;
        this.selected = undefined;
        this.accessibleDescriptionDefault = null;
        this.accessibleDescription = this
            .accessibleDescriptionDefault
            ? getLocaleString(this.accessibleDescriptionDefault)
            : null;
        this.accessibleLabelledBy = undefined;
    }
    /**
     * Component lifecycle events.
     */
    connectedCallback() {
        connectLanguageChangeObserver(this, { prop: "accessibleDescription", defaults: "accessibleDescriptionDefault" });
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        if (Build.isServer) {
            return;
        }
        return (h("div", { role: "tabpanel", class: { "duet-tab": true, selected: this.selected }, "aria-description": this.accessibleDescription, "aria-labelledby": this.accessibleLabelledBy }, h("slot", null)));
    }
    static get is() { return "duet-tab"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-tab.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-tab.css"]
        };
    }
    static get properties() {
        return {
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label for the tab."
                },
                "attribute": "label",
                "reflect": true
            },
            "caption": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Caption for the tab."
                },
                "attribute": "caption",
                "reflect": true
            },
            "selected": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Selected state of the tab."
                },
                "attribute": "selected",
                "reflect": true
            },
            "accessibleDescriptionDefault": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject | string",
                    "resolved": "DuetLangObject | string",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "null"
                        }],
                    "text": "Defaults for AccessibleLabel"
                },
                "attribute": "accessible-description-defaults",
                "reflect": false,
                "defaultValue": "null"
            },
            "accessibleDescription": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "null"
                        }],
                    "text": "String for AccessibleLabel"
                },
                "attribute": "accessible-description",
                "reflect": false,
                "defaultValue": "this\n    .accessibleDescriptionDefault\n    ? getLocaleString(this.accessibleDescriptionDefault)\n    : null"
            },
            "accessibleLabelledBy": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "String of id's that indicate alternative labels elements"
                },
                "attribute": "accessible-labelled-by",
                "reflect": false,
                "defaultValue": "undefined"
            }
        };
    }
    static get elementRef() { return "element"; }
}
