// @duetds/components/lib/esm/duet-tooltip.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host, g as getElement } from './index-909a963e.js';
import { m as media_query_medium, a as media_query_large } from './tokens-04106a43.js';
import { c as computePosition, f as flip, s as shift, a as arrow } from './floating-ui.dom.esm-efad5c89.js';
import { i as inheritGlobalTheme } from './themeable-component-724c0f7e.js';
import { c as createID } from './create-id-149a1b6d.js';
import { f as isEscapeKey } from './keyboard-utils-eba88365.js';
import { a as getLanguage, g as getLocaleString, c as connectLanguageChangeObserver, d as disconnectLanguageChangeObserver } from './language-utils-00ca4c55.js';
import './string-utils-ca039233.js';

const duetTooltipCss = "*.sc-duet-tooltip,*.sc-duet-tooltip::after,*.sc-duet-tooltip::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.sc-duet-tooltip-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;z-index:200;display:inline-flex;width:auto;margin-top:-16px !important;margin-left:-3px !important;-webkit-tap-highlight-color:transparent}duet-tooltip-button.sc-duet-tooltip{position:relative}duet-tooltip-popup.sc-duet-tooltip{position:absolute}";

const DuetTooltip = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /**
         * Own Properties.
         */
        this.mql = [
            window.matchMedia(media_query_medium.replace(/'/g, "")),
            window.matchMedia(media_query_large.replace(/'/g, "")),
        ];
        this.tooltipId = createID("DuetTooltip");
        this.aboutString = {
            fi: "aiheesta",
            en: "about",
            sv: "om",
        };
        /**
         * Local methods
         */
        this.openTooltip = () => {
            this.active = true;
            this.focusTarget = "tooltip";
        };
        this.closeTooltip = (moveFocusToButton = false) => {
            this.active = false;
            if (moveFocusToButton) {
                this.focusTarget = "trigger";
            }
            setTimeout(() => {
                this.updatePlacement();
            }, 200);
        };
        this.toggleOpen = () => {
            this.active ? this.closeTooltip() : this.openTooltip();
        };
        this.updatePlacement = async () => {
            var _a, _b;
            let preferRight = true;
            if (this.breakpoint === "large" && !this.mql[1].matches) {
                preferRight = false;
            }
            else if (this.breakpoint === "medium" && !this.mql[0].matches) {
                preferRight = false;
            }
            let placement = preferRight ? "top-start" : "top-end";
            if (this.direction === "left") {
                placement = "top-end";
            }
            else if (this.direction === "right") {
                placement = "top-start";
            }
            const position = await computePosition(this.buttonEl, this.popupElement, {
                placement,
                middleware: [flip(), shift({ padding: 5 }), arrow({ element: this.arrowEl })],
            });
            this.tooltipY = position.y;
            this.tooltipX = position.x;
            this.placement = position.placement;
            this.arrowY = (_a = position.middlewareData.arrow) === null || _a === void 0 ? void 0 : _a.y;
            this.arrowX = (_b = position.middlewareData.arrow) === null || _b === void 0 ? void 0 : _b.x;
        };
        this.placement = "bottom";
        this.tooltipY = undefined;
        this.tooltipX = undefined;
        this.arrowX = undefined;
        this.arrowY = undefined;
        this.theme = "";
        this.language = getLanguage();
        this.accessibleLabelDefault = {
            fi: "Näytä lisätietoja",
            en: "Show more information",
            sv: "Visa mer tilläggsuppgifter",
        };
        this.accessibleLabel = getLocaleString(this.accessibleLabelDefault, this.language);
        this.accessibleAboutLabel = undefined;
        this.accessibleInputLabel = undefined;
        this.accessibleCloseLabelDefault = {
            fi: "Sulje lisätiedot",
            en: "Close details",
            sv: "Stäng tilläggsuppgifter",
        };
        this.accessibleCloseLabel = getLocaleString(this.accessibleCloseLabelDefault, this.language);
        this.label = "";
        this.breakpoint = "medium";
        this.direction = "auto";
        this.positionVertical = "0";
        this.positionHorizontal = "0";
        this.active = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    async componentDidLoad() {
        this.arrowEl = await this.popupElement.getArrowElement();
        this.contentEl = await this.popupElement.getContentElement();
        this.updatePlacement();
        if (this.active) {
            this.openTooltip();
        }
        this.intersectionObserver = new IntersectionObserver(this.updatePlacement, { root: document.body });
        this.intersectionObserver.observe(this.element);
    }
    componentDidUpdate() {
        switch (this.focusTarget) {
            case "tooltip":
                // delay focus so button is correctly announced as "expanded"
                setTimeout(() => this.contentEl.focus({ preventScroll: true }));
                break;
            case "trigger":
                // delay focus so button is correctly announced as "collapsed"
                setTimeout(() => this.buttonEl.setFocus({ preventScroll: true }));
                break;
        }
        this.focusTarget = null;
    }
    connectedCallback() {
        connectLanguageChangeObserver(this, [
            { prop: "accessibleLabel", defaults: "accessibleLabelDefault" },
            { prop: "accessibleCloseLabel", defaults: "accessibleCloseLabelDefault" },
        ]);
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
    }
    /**
     * Component event handling.
     */
    onScroll() {
        clearTimeout(this.scrollTimer);
        this.scrollTimer = setTimeout(() => {
            if (this.active) {
                this.updatePlacement();
            }
        }, 100);
    }
    handleWindowResize() {
        clearTimeout(this.resizeTimer);
        this.resizeTimer = setTimeout(() => {
            this.updatePlacement();
        }, 100);
    }
    handleDocumentClick(e) {
        if (!this.active) {
            return;
        }
        const path = e.composedPath();
        const isClickOutside = path.every(el => el !== this.element && el !== this.contentEl);
        if (isClickOutside) {
            this.closeTooltip();
        }
    }
    handleKeyDown(e) {
        if (this.active && isEscapeKey(e)) {
            this.closeTooltip(true);
        }
    }
    handlePopupClose() {
        if (this.active) {
            this.closeTooltip(true);
        }
    }
    async handleActiveChange(newActiveValue) {
        if (newActiveValue) {
            setTimeout(() => this.updatePlacement(), 10);
        }
        else {
            setTimeout(() => this.updatePlacement(), 300);
        }
    }
    /**
     * render() function.
     * Always the last one in the class.
     */
    render() {
        const tooltipStyles = {
            transform: `translateY(${this.positionVertical}) translateX(${this.positionHorizontal})`,
            // This is needed to that the active tooltip stacks on
            // top of other tooltips and the header.
            zIndex: this.active ? "201" : "",
        };
        const popupContainerStyles = {
            position: "relative",
            overflow: this.active ? "visible" : "hidden",
        };
        const positionStyles = {
            top: `${this.tooltipY}px`,
            left: `${this.tooltipX}px`,
            display: "block",
            "pointer-events": this.active ? "auto" : "none",
        };
        const arrowStyles = {
            top: this.arrowY ? `${this.arrowY}px` : undefined,
            left: this.arrowX ? `${this.arrowX}px` : undefined,
        };
        return (h(Host, { style: tooltipStyles }, h("duet-tooltip-button", { ref: element => (this.buttonEl = element), active: this.active, tooltipId: this.tooltipId, label: this.label, accessibleLabel: (this.accessibleInputLabel && `${this.accessibleInputLabel}, ${this.accessibleLabel}`) ||
                (this.accessibleAboutLabel &&
                    `${this.accessibleLabel} ${this.aboutString[this.language]} ${this.accessibleAboutLabel}`) ||
                this.accessibleLabel, onClick: this.toggleOpen, theme: this.theme }), h("div", { style: popupContainerStyles }, h("duet-tooltip-popup", { ref: element => (this.popupElement = element), active: this.active, tooltipId: this.tooltipId, arrowStyles: arrowStyles, style: positionStyles, placement: this.placement, accessibleCloseLabel: this.accessibleCloseLabel, theme: this.theme }, h("slot", null)))));
    }
    get element() { return getElement(this); }
    static get watchers() { return {
        "active": ["handleActiveChange"]
    }; }
};
DuetTooltip.style = duetTooltipCss;

export { DuetTooltip as duet_tooltip };
