// @duetds/components/lib/collection/components/duet-file-chooser/duet-file-chooser.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common";
export class DuetUploadItem {
    constructor() {
        this.onChange = (ev) => {
            ev.preventDefault();
            ev.stopPropagation();
            this.duetChange.emit({
                originalEvent: ev,
                data: { files: this.nativeInput.files },
                component: "duet-file-chooser",
            });
        };
        this.theme = "";
        this.multiple = false;
        this.accept = undefined;
    }
    /**
     * Opens file choose dialog.
     */
    async open() {
        var _a;
        (_a = this.nativeInput) === null || _a === void 0 ? void 0 : _a.click();
    }
    /**
     * Returns actually selected files in file chooser.
     */
    async getFiles() {
        return this.nativeInput.files;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    render() {
        return (h(Host, null, h("duet-visually-hidden", { "aria-hidden": "true" }, h("input", { ref: input => {
                this.nativeInput = input;
            }, type: "file", class: {
                "duet-upload": true,
            }, "aria-hidden": "true", tabIndex: -1, onChange: this.onChange, multiple: this.multiple, accept: this.accept }))));
    }
    static get is() { return "duet-file-chooser"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-file-chooser.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-file-chooser.css"]
        };
    }
    static get properties() {
        return {
            "theme": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetTheme",
                    "resolved": "\"\" | \"default\" | \"turva\"",
                    "references": {
                        "DuetTheme": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetTheme"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "theme."
                },
                "attribute": "theme",
                "reflect": false,
                "defaultValue": "\"\""
            },
            "multiple": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Use multiple to allow the user to select multiple files when uploading"
                },
                "attribute": "multiple",
                "reflect": false,
                "defaultValue": "false"
            },
            "accept": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The accept attribute takes as its value a comma-separated list of one or more file types,\nor unique file type specifiers, describing which file types to allow."
                },
                "attribute": "accept",
                "reflect": false,
                "defaultValue": "undefined"
            }
        };
    }
    static get events() {
        return [{
                "method": "duetChange",
                "name": "duetChange",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Emitted when the value has changed."
                },
                "complexType": {
                    "original": "DuetFileChooserEvent",
                    "resolved": "{ originalEvent?: Event; data?: Record<string, any>; component: \"duet-file-chooser\"; }",
                    "references": {
                        "DuetFileChooserEvent": {
                            "location": "local",
                            "path": "/Users/mgibas/workspace/duet/packages/components/src/components/duet-file-chooser/duet-file-chooser.tsx",
                            "id": "src/components/duet-file-chooser/duet-file-chooser.tsx::DuetFileChooserEvent"
                        }
                    }
                }
            }];
    }
    static get methods() {
        return {
            "open": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Opens file choose dialog.",
                    "tags": []
                }
            },
            "getFiles": {
                "complexType": {
                    "signature": "() => Promise<FileList>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "FileList": {
                            "location": "global",
                            "id": "global::FileList"
                        }
                    },
                    "return": "Promise<FileList>"
                },
                "docs": {
                    "text": "Returns actually selected files in file chooser.",
                    "tags": []
                }
            }
        };
    }
}
