// @duetds/components/lib/collection/components/duet-toolbar-link/duet-toolbar-link.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetToolbarLink {
    constructor() {
        this.theme = "";
        this.variation = "default";
        this.active = false;
        this.href = "#";
        this.icon = undefined;
        this.accessibleLabel = undefined;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        const variation = this.variation === "negative" ? "negative" : "white";
        return (h(Host, null, h("a", { class: {
                "duet-toolbar-link": true,
                "duet-theme-turva": this.theme === "turva",
                active: this.active,
                [variation]: true,
            }, href: this.href, "aria-current": this.active ? "true" : "false", "aria-label": this.accessibleLabel }, this.icon && (h("duet-icon", { size: "x-small", name: this.icon, margin: "none", theme: this.theme, color: "currentColor" })), h("slot", null))));
    }
    static get is() { return "duet-toolbar-link"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-toolbar-link.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-toolbar-link.css"]
        };
    }
    static get properties() {
        return {
            "theme": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetTheme",
                    "resolved": "\"\" | \"default\" | \"turva\"",
                    "references": {
                        "DuetTheme": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetTheme"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Theme."
                },
                "attribute": "theme",
                "reflect": false,
                "defaultValue": "\"\""
            },
            "variation": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "\"default\" | \"negative\"",
                    "resolved": "\"default\" | \"negative\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Variation."
                },
                "attribute": "variation",
                "reflect": false,
                "defaultValue": "\"default\""
            },
            "active": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Active"
                },
                "attribute": "active",
                "reflect": false,
                "defaultValue": "false"
            },
            "href": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Href"
                },
                "attribute": "href",
                "reflect": false,
                "defaultValue": "\"#\""
            },
            "icon": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Icon"
                },
                "attribute": "icon",
                "reflect": true,
                "defaultValue": "undefined"
            },
            "accessibleLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Adds accessible label for the link that is only shown for screen readers.\nTypically, this label text replaces the visible text on the link for users\nwho use assistive technology."
                },
                "attribute": "accessible-label",
                "reflect": false
            }
        };
    }
    static get elementRef() { return "element"; }
}
