// @duetds/components/lib/collection/components/duet-shaped-image/duet-shaped-image.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetShapedImage {
    constructor() {
        this.theme = "";
        this.accessibleLabel = undefined;
        this.margin = "auto";
        this.src = undefined;
        this.rotation = 7;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
    }
    /**
     * render() function
     * Always the last one of the class.
     */
    render() {
        const style = {
            "--rotation": `${this.rotation}deg`,
        };
        return (h(Host, { class: { "duet-theme-turva": this.theme === "turva", "duet-m-0": this.margin === "none" } }, h("span", { class: {
                "shaped-image": true,
            }, role: "img", "aria-label": this.accessibleLabel, style: style }, h("span", { class: "shaped-image-mask", role: "presentation" }, h("span", { class: "shaped-image-img", role: "presentation", style: { backgroundImage: `url("${this.src}")` } })))));
    }
    static get is() { return "duet-shaped-image"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-shaped-image.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-shaped-image.css"]
        };
    }
    static get properties() {
        return {
            "theme": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetTheme",
                    "resolved": "\"\" | \"default\" | \"turva\"",
                    "references": {
                        "DuetTheme": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetTheme"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Theme of the component."
                },
                "attribute": "theme",
                "reflect": false,
                "defaultValue": "\"\""
            },
            "accessibleLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Adds accessible label for the image that is only shown for screen readers."
                },
                "attribute": "accessible-label",
                "reflect": false
            },
            "margin": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetMargin",
                    "resolved": "\"auto\" | \"none\"",
                    "references": {
                        "DuetMargin": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetMargin"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the margin of the component."
                },
                "attribute": "margin",
                "reflect": false,
                "defaultValue": "\"auto\""
            },
            "src": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "A URL from which to load an image."
                },
                "attribute": "src",
                "reflect": false
            },
            "rotation": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "7 | -7",
                    "resolved": "-7 | 7",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "\"7\""
                        }],
                    "text": "Rotation. Note: in 'turva' theme there's no rotation."
                },
                "attribute": "rotation",
                "reflect": false,
                "defaultValue": "7"
            }
        };
    }
}
