// @duetds/components/lib/collection/components/duet-link/duet-link.js
/*!
 * Built with Duet Design System
 */
import chevronIcon from "@duetds/icons/lib/assets/action-arrow-right";
import newWindowIcon from "@duetds/icons/lib/assets/action-new-window-small";
import { Build, Fragment, h, Host } from "@stencil/core";
import { DuetStringsExternalDefaults } from "../../common-strings";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { connectLanguageChangeObserver, disconnectLanguageChangeObserver, getLanguage, getLocaleString, } from "../../utils/language-utils";
import { breakpointToToken } from "../../utils/token-utils";
/**
 * @part duet-link - piercing selector for styling the link
 * @part duet-link-content - piercing selector for styling the link content wrapper
 * @part block-chevron - piercing selector for styling the block chevron
 */
export class DuetLink {
    constructor() {
        this.variation = "default";
        this.accessibleLabel = undefined;
        this.language = getLanguage();
        this.accessibleLabelExternalDefaults = DuetStringsExternalDefaults;
        this.accessibleLabelExternal = getLocaleString(this.accessibleLabelExternalDefaults);
        this.identifier = undefined;
        this.theme = "";
        this.url = "#";
        this.external = false;
        this.tabToLink = false;
        this.icon = undefined;
        this.iconColor = "currentColor";
        this.iconResponsive = false;
        this.padding = "auto";
        this.isSmallScreen = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
        this.initMediaQuery();
    }
    connectedCallback() {
        connectLanguageChangeObserver(this, {
            prop: "accessibleLabelExternal",
            defaults: "accessibleLabelExternalDefaults",
        });
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
        this.destroyMediaQuery();
    }
    /**
     * Sets focus on the specified `duet-link`. Use this method instead of the global
     * `link.focus()`.
     */
    async setFocus(options) {
        this.nativeLink.focus(options);
    }
    initMediaQuery() {
        if (Build.isServer) {
            return;
        }
        this.screenTestBound = this.screenTest.bind(this);
        this.mediaQuery = window.matchMedia(breakpointToToken["xx-small"]);
        this.isSmallScreen = this.mediaQuery.matches;
        this.mediaQuery.addEventListener("change", this.screenTestBound);
    }
    destroyMediaQuery() {
        if (Build.isServer || !this.mediaQuery) {
            return;
        }
        this.mediaQuery.removeEventListener("change", this.screenTestBound);
    }
    screenTest(event) {
        this.isSmallScreen = event.matches;
    }
    getAccessibleLabel() {
        if (this.accessibleLabel) {
            return this.accessibleLabel;
        }
        if (!this.element.innerHTML) {
            return "";
        }
        const capitalized = this.element.innerHTML[0].toUpperCase() + this.element.innerHTML.slice(1);
        return `${capitalized}. ${this.external ? `${this.accessibleLabelExternal}.` : ""}`;
    }
    renderContent() {
        return (h(Fragment, null, h("span", { class: "duet-link-content", part: "duet-link-content" }, h("slot", null)), Build.isBrowser && this.external && (h(Fragment, null, h("span", { class: "duet-link-external" }, "\u00A0", h("duet-icon", { icon: newWindowIcon.svg, size: "xx-small", margin: "none", color: "currentColor" }))))));
    }
    renderGrid() {
        return (h(Fragment, null, this.icon && (h("span", { class: "block__icon" }, h("duet-icon", { size: this.iconResponsive ? (this.isSmallScreen ? "medium" : "large") : "x-small", name: this.icon, margin: "none", color: this.iconColor }))), h("span", { class: "block__content" }, this.renderContent()), h("span", { class: "block__chevron", part: "block-chevron" }, h("duet-icon", { class: "block__chevron-icon", icon: chevronIcon.svg, size: "x-small", margin: "none", color: "primary-dark" }))));
    }
    /**
     * render() function
     * Always the last one in the class.
     */
    render() {
        // Notice that the external icon needs to be places right after the last word of the link.
        // This is why we use &nbps; to keep everything on the same line.
        return (h(Host, { class: {
                "duet-link-block": this.variation === "block",
            } }, h("a", { href: this.url, class: {
                "duet-link": true,
                "duet-link-is-external": this.external,
                "duet-link-has-icon": !!this.icon,
                "duet-tab-to-link": this.tabToLink,
                "duet-theme-turva": this.theme === "turva",
                "duet-p-0": this.padding === "none",
                [this.padding]: true,
                [this.variation]: true,
            }, target: this.external ? "_blank" : "_self", "aria-label": this.getAccessibleLabel(), id: this.identifier, ref: el => (this.nativeLink = el), part: "duet-link", rel: this.external ? "noopener" : undefined }, this.variation === "block" ? this.renderGrid() : this.renderContent())));
    }
    static get is() { return "duet-link"; }
    static get encapsulation() { return "shadow"; }
    static get delegatesFocus() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-link.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-link.css"]
        };
    }
    static get properties() {
        return {
            "variation": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetLinkVariation",
                    "resolved": "\"block\" | \"default\"",
                    "references": {
                        "DuetLinkVariation": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetLinkVariation"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Style variation of the link."
                },
                "attribute": "variation",
                "reflect": false,
                "defaultValue": "\"default\""
            },
            "accessibleLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Adds accessible label for the link that is only shown for screen readers.\nTypically, this label text replaces the visible text on the link for users\nwho use assistive technology."
                },
                "attribute": "accessible-label",
                "reflect": false
            },
            "language": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetLanguage",
                    "resolved": "\"en\" | \"fi\" | \"sv\"",
                    "references": {
                        "DuetLanguage": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetLanguage"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "deprecated",
                            "text": "this is now handled via the html lang tag, and is no longer used - kept to avoid breaking changes and ease unit testing"
                        }, {
                            "name": "default",
                            "text": "\"fi\""
                        }],
                    "text": "The currently active language. This setting changes the external link\naccessible label to match the chosen language."
                },
                "attribute": "language",
                "reflect": true,
                "defaultValue": "getLanguage()"
            },
            "accessibleLabelExternalDefaults": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject | string",
                    "resolved": "DuetLangObject | string",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "{fi: \"Avautuu uuteen ikkunaan\",sv: \"\u00D6ppnas i nytt f\u00F6nster\",en: \"Opens in a new window\"}"
                        }],
                    "text": "Defaults for accessibleLabelExternal"
                },
                "attribute": "accessible-label-external-default",
                "reflect": false,
                "defaultValue": "DuetStringsExternalDefaults"
            },
            "accessibleLabelExternal": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "default",
                            "text": "{fi: \"Avautuu uuteen ikkunaan\",sv: \"\u00D6ppnas i nytt f\u00F6nster\",en: \"Opens in a new window\"}"
                        }],
                    "text": "Adds accessible label for tooltip that is shown in external link (url & external have both been set)"
                },
                "attribute": "accessible-label-external",
                "reflect": false,
                "defaultValue": "getLocaleString(this.accessibleLabelExternalDefaults)"
            },
            "identifier": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Adds a unique identifier for the link. Please note that with this particular\ncomponent this id is added inside Shadow DOM. If you need an id on the html\nelement, use regular id attribute instead."
                },
                "attribute": "identifier",
                "reflect": false
            },
            "theme": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetTheme",
                    "resolved": "\"\" | \"default\" | \"turva\"",
                    "references": {
                        "DuetTheme": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetTheme"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Theme of the button."
                },
                "attribute": "theme",
                "reflect": false,
                "defaultValue": "\"\""
            },
            "url": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "A destination to link to, rendered in the href attribute of a link."
                },
                "attribute": "url",
                "reflect": true,
                "defaultValue": "\"#\""
            },
            "external": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Forces URL to open in a new browser tab. Used together with URL prop."
                },
                "attribute": "external",
                "reflect": false,
                "defaultValue": "false"
            },
            "tabToLink": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Initially visually hidden link, which is shown on focus."
                },
                "attribute": "tab-to-link",
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetIconName",
                    "resolved": "string",
                    "references": {
                        "DuetIconName": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetIconName"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Icon to display on the left side of the ink text (only with \"block\" variation)"
                },
                "attribute": "icon",
                "reflect": false
            },
            "iconColor": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetIconColor",
                    "resolved": "string",
                    "references": {
                        "DuetIconColor": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetIconColor"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Color of the icon to display (only with \"block\" variation)"
                },
                "attribute": "icon-color",
                "reflect": false,
                "defaultValue": "\"currentColor\""
            },
            "iconResponsive": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the icon changes its size responsively (only with \"block\" variation)"
                },
                "attribute": "icon-responsive",
                "reflect": false,
                "defaultValue": "false"
            },
            "padding": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetPadding",
                    "resolved": "\"auto\" | \"none\"",
                    "references": {
                        "DuetPadding": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetPadding"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the padding of the component (only with \"block\" variation)"
                },
                "attribute": "padding",
                "reflect": false,
                "defaultValue": "\"auto\""
            }
        };
    }
    static get states() {
        return {
            "isSmallScreen": {}
        };
    }
    static get methods() {
        return {
            "setFocus": {
                "complexType": {
                    "signature": "(options?: FocusOptions) => Promise<void>",
                    "parameters": [{
                            "name": "options",
                            "type": "FocusOptions",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "FocusOptions": {
                            "location": "global",
                            "id": "global::FocusOptions"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Sets focus on the specified `duet-link`. Use this method instead of the global\n`link.focus()`.",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "element"; }
}
