// @duetds/components/lib/collection/utils/axe-core-utils.js
/*!
 * Built with Duet Design System
 */
import { createPage } from "./test-utils";
const axeCorePath = require.resolve("axe-core");
export const options = {
    rules: {
        "document-title": { enabled: false },
        "html-has-lang": { enabled: false },
        "landmark-one-main": { enabled: false },
        "page-has-heading-one": { enabled: false },
        region: { enabled: false },
    },
};
export function formatViolations(violations = []) {
    return violations.map(({ id, impact, description }) => ({
        id,
        impact,
        description,
    }));
}
export function mergeRules(axeRunOptions = {}, rules) {
    return rules
        ? {
            rules: Object.assign(Object.assign({}, axeRunOptions.rules), rules),
        }
        : axeRunOptions;
}
export async function testWithAxe(markup, axeRunOptions = {}) {
    const page = await createPage(markup);
    await page.addScriptTag({ path: axeCorePath });
    await new Promise(process.nextTick);
    return page.evaluate(async (options) => {
        // @ts-ignore
        return await axe.run(options);
    }, axeRunOptions);
}
