// @duetds/components/lib/collection/components/duet-show-more/duet-show-more.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { createID } from "../../utils/create-id";
import { isKeyboardClick } from "../../utils/keyboard-utils";
import { connectLanguageChangeObserver, disconnectLanguageChangeObserver, getLanguage, getLocaleString, } from "../../utils/language-utils";
export class DuetShowMore {
    constructor() {
        this.contentId = createID("DuetShowMore");
        this.buttonId = createID("DuetShowMoreButton");
        this.handleKeyDown = (e) => {
            if (isKeyboardClick(e)) {
                this.handleClick(e);
            }
        };
        this.handleClick = (e) => {
            e.preventDefault();
            const open = !this.open;
            this.open = open;
            if (open) {
                setTimeout(() => {
                    this.contentElement.focus();
                }, 100);
                setTimeout(() => {
                    this.buttonAriaOpen = open;
                }, 500);
            }
            else {
                this.buttonAriaOpen = open;
            }
            this.duetToggle.emit({
                component: "duet-show-more",
                originalEvent: e,
            });
        };
        this.buttonAriaOpen = false;
        this.theme = "";
        this.language = getLanguage();
        this.showMoreLabelDefaults = {
            fi: "Näytä lisää",
            sv: "Visa mer",
            en: "Show more",
        };
        this.showMoreLabel = getLocaleString(this.showMoreLabelDefaults);
        this.showLessLabelDefaults = {
            fi: "Näytä vähemmän",
            sv: "Visa mindre",
            en: "Show less",
        };
        this.showLessLabel = getLocaleString(this.showLessLabelDefaults);
        this.showMoreAccessibleLabelDefaults = {
            fi: "Näytä lisää tästä aiheesta",
            sv: "Visa mer om detta ämne",
            en: "Show more about this subject",
        };
        this.showMoreAccessibleLabel = getLocaleString(this.showMoreAccessibleLabelDefaults);
        this.showLessAccessibleLabelDefaults = {
            fi: "Näytä vähemmän tästä aiheesta",
            sv: "Visa mindre om detta ämne",
            en: "Show less about this subject",
        };
        this.showLessAccessibleLabel = getLocaleString(this.showLessAccessibleLabelDefaults);
        this.margin = "auto";
        this.buttonMargin = "auto";
        this.buttonPadding = "auto";
        this.open = false;
    }
    /**
     * Component lifecycle events.
     */
    componentWillLoad() {
        inheritGlobalTheme(this);
        this.buttonAriaOpen = this.open;
    }
    connectedCallback() {
        connectLanguageChangeObserver(this, [
            { prop: "showMoreLabel", defaults: "showMoreLabelDefaults" },
            { prop: "showLessLabel", defaults: "showLessLabelDefaults" },
            { prop: "showMoreAccessibleLabel", defaults: "showMoreAccessibleLabelDefaults" },
            { prop: "showLessAccessibleLabel", defaults: "showLessAccessibleLabelDefaults" },
        ]);
    }
    disconnectedCallback() {
        disconnectLanguageChangeObserver(this);
    }
    /**
     * render() function
     * Always the last one of the class.
     */
    render() {
        return (h(Host, null, h("div", { class: {
                "duet-show-more": true,
                "duet-theme-turva": this.theme === "turva",
                "duet-m-0": this.margin === "none",
                open: this.open,
            } }, h("div", { ref: ref => (this.contentElement = ref), class: { content: true }, id: this.contentId, tabindex: this.open ? "0" : "" }, h("div", { class: { "slotted-content": true } }, h("slot", null))), h("duet-button", { id: this.buttonId, variation: "plain", "accessible-label": this.buttonAriaOpen ? this.showLessAccessibleLabel : this.showMoreAccessibleLabel, "accessible-expanded": this.buttonAriaOpen ? "true" : "false", "accessible-controls": this.contentId, onClick: this.handleClick, onKeyDown: this.handleKeyDown, "icon-right": true, icon: this.open ? "action-arrow-up-small" : "action-arrow-down-small", "icon-size": "medium-small", size: "small", margin: this.buttonMargin, padding: this.buttonPadding, theme: this.theme === "turva" ? "turva" : "default" }, this.open ? this.showLessLabel : this.showMoreLabel))));
    }
    static get is() { return "duet-show-more"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["duet-show-more.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["duet-show-more.css"]
        };
    }
    static get properties() {
        return {
            "theme": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetTheme",
                    "resolved": "\"\" | \"default\" | \"turva\"",
                    "references": {
                        "DuetTheme": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetTheme"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Theme."
                },
                "attribute": "theme",
                "reflect": false,
                "defaultValue": "\"\""
            },
            "language": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "DuetLanguage",
                    "resolved": "\"en\" | \"fi\" | \"sv\"",
                    "references": {
                        "DuetLanguage": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetLanguage"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "deprecated",
                            "text": "this is now handled via the html lang tag, and is no longer used - kept to avoid breaking changes and ease unit testing"
                        }, {
                            "name": "default",
                            "text": "\"fi\""
                        }],
                    "text": "The currently active language. This setting also changes the logo to match\nthe chosen language."
                },
                "attribute": "language",
                "reflect": true,
                "defaultValue": "getLanguage()"
            },
            "showMoreLabelDefaults": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject",
                    "resolved": "DuetLangObject",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Default translations for show more label."
                },
                "defaultValue": "{\n    fi: \"N\u00E4yt\u00E4 lis\u00E4\u00E4\",\n    sv: \"Visa mer\",\n    en: \"Show more\",\n  }"
            },
            "showMoreLabel": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show more label. Defaults to what is in showMoreLabelDefaults prop."
                },
                "attribute": "show-more-label",
                "reflect": true,
                "defaultValue": "getLocaleString(this.showMoreLabelDefaults)"
            },
            "showLessLabelDefaults": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject",
                    "resolved": "DuetLangObject",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Default translations for show less label."
                },
                "defaultValue": "{\n    fi: \"N\u00E4yt\u00E4 v\u00E4hemm\u00E4n\",\n    sv: \"Visa mindre\",\n    en: \"Show less\",\n  }"
            },
            "showLessLabel": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show less label. Defaults to what is in showLessLabelDefaults prop."
                },
                "attribute": "show-less-label",
                "reflect": true,
                "defaultValue": "getLocaleString(this.showLessLabelDefaults)"
            },
            "showMoreAccessibleLabelDefaults": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject",
                    "resolved": "DuetLangObject",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Default translations for show more accessible label."
                },
                "defaultValue": "{\n    fi: \"N\u00E4yt\u00E4 lis\u00E4\u00E4 t\u00E4st\u00E4 aiheesta\",\n    sv: \"Visa mer om detta \u00E4mne\",\n    en: \"Show more about this subject\",\n  }"
            },
            "showMoreAccessibleLabel": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show more accessible label. Defaults to what is in showMoreAccessibleLabelDefaults prop."
                },
                "attribute": "show-more-accessible-label",
                "reflect": true,
                "defaultValue": "getLocaleString(\n    this.showMoreAccessibleLabelDefaults\n  )"
            },
            "showLessAccessibleLabelDefaults": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "DuetLangObject",
                    "resolved": "DuetLangObject",
                    "references": {
                        "DuetLangObject": {
                            "location": "import",
                            "path": "../../utils/language-utils",
                            "id": "src/utils/language-utils.ts::DuetLangObject"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Default translations for show less accessible label."
                },
                "defaultValue": "{\n    fi: \"N\u00E4yt\u00E4 v\u00E4hemm\u00E4n t\u00E4st\u00E4 aiheesta\",\n    sv: \"Visa mindre om detta \u00E4mne\",\n    en: \"Show less about this subject\",\n  }"
            },
            "showLessAccessibleLabel": {
                "type": "string",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show less accessible label. Defaults to what is in showLessAccessibleLabelDefaults prop."
                },
                "attribute": "show-less-accessible-label",
                "reflect": true,
                "defaultValue": "getLocaleString(\n    this.showLessAccessibleLabelDefaults\n  )"
            },
            "margin": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetMargin",
                    "resolved": "\"auto\" | \"none\"",
                    "references": {
                        "DuetMargin": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetMargin"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the margin of the component."
                },
                "attribute": "margin",
                "reflect": false,
                "defaultValue": "\"auto\""
            },
            "buttonMargin": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetMargin",
                    "resolved": "\"auto\" | \"none\"",
                    "references": {
                        "DuetMargin": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetMargin"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the margin of the button component inside of show more."
                },
                "attribute": "button-margin",
                "reflect": false,
                "defaultValue": "\"auto\""
            },
            "buttonPadding": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "DuetPadding",
                    "resolved": "\"auto\" | \"none\"",
                    "references": {
                        "DuetPadding": {
                            "location": "import",
                            "path": "../../common-types",
                            "id": "src/common-types.d.ts::DuetPadding"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls the padding of the button component inside of show more."
                },
                "attribute": "button-padding",
                "reflect": false,
                "defaultValue": "\"auto\""
            },
            "open": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Is the component expanded or collapsed."
                },
                "attribute": "open",
                "reflect": true,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "buttonAriaOpen": {}
        };
    }
    static get events() {
        return [{
                "method": "duetToggle",
                "name": "duetToggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when opened/closed"
                },
                "complexType": {
                    "original": "DuetShowMoreToggleEvent",
                    "resolved": "{ component: \"duet-show-more\"; originalEvent: KeyboardEvent | MouseEvent; }",
                    "references": {
                        "DuetShowMoreToggleEvent": {
                            "location": "local",
                            "path": "/Users/mgibas/workspace/duet/packages/components/src/components/duet-show-more/duet-show-more.tsx",
                            "id": "src/components/duet-show-more/duet-show-more.tsx::DuetShowMoreToggleEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "element"; }
}
